<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title></title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:30px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">




</div>


<!-- rmd lab header -->
<p><br></p>
<div id="precourse" class="section level1 tabset tabset-fade">
<h1>Precourse</h1>
<p>This workshop is aimed towards biologists, researchers, computer scientists or data analysts planning to run, analyse and interpret an RNA-Seq experiment. Basic knowledge of working on the Unix/Linux command line is expected.</p>
<div id="preparation" class="section level2">
<h2>Preparation</h2>
<p>These are steps to be completed before the workshop.</p>
<div id="uppmax-id" class="section level3">
<h3>Uppmax ID</h3>
<p>A remote computing cluster (UPPMAX) will be use for data analyses. A SUPR/SNIC account is needed to use UPPMAX resources.</p>
<p>If you do not already have one, create an account at <a href="https://supr.snic.se/">https://supr.snic.se/</a>.</p>
<p><img src="data/common/supr-login.png" class="fancyimage"></p>
<p>Log in to <a href="https://supr.snic.se/">SUPR/SNIC</a> and request membership to the project ID <strong>g2020006</strong>.</p>
<p><img src="data/common/supr-request.png" class="fancyimage"></p>
<p>Once you are accepted to the project, you should see that project listed under your active projects.</p>
<p><img src="data/common/supr-projects.png" class="fancyimage"></p>
</div>
<div id="ssh-sftp-access" class="section level3">
<h3>SSH &amp; SFTP access</h3>
<div class="boxy boxy-lightbulb">
<p>You need a program to connect to a remote cluster (UPPMAX). Linux and Mac users already have terminal on their systems.</p>
<p><a href="http://mobaxterm.mobatek.net"><img src="data/common/mobaxterm.png" /></a></p>
<p>If you are on a Windows system, we recommend <a href="http://mobaxterm.mobatek.net">MobaXterm</a>. It is recommended that you INSTALL the program and not use the portable version. MobaXterm also has an integrated SFTP file browser.</p>
<p><a href="https://www.xquartz.org/"><img src="data/common/xquartz.png" /></a></p>
<p>Mac users will need to download and install <a href="https://www.xquartz.org/">XQuartz</a> for X11 forwarding. <em>ie</em>; to forward remotely opened windows to local machine.</p>
</div>
<p>Type <code>ssh -Y user@rackham.uppmax.uu.se</code> in the terminal and then enter your password. The password will not be visible as you type.</p>
<div class="boxy boxy-lightbulb">
<p>If you need to transfer data between UPPMAX and your computer, you can use tools SCP or SFTP through the terminal. Windows users can use the SFTP browser available with MobaXterm.</p>
<p><a href="https://filezilla-project.org/"><img src="data/common/filezilla.png" /></a></p>
<p>For all users, if you prefer a GUI to upload and download files from UPPMAX, we recommend installing <a href="https://filezilla-project.org/">FileZilla</a>.</p>
</div>
</div>
<div id="course-directory-access" class="section level3">
<h3>Course directory access</h3>
<div class="boxy boxy-lightbulb">
<p>For this step, you will have to use the terminal a bit. You can get started by following <strong>Tutorial One</strong> at this link <em><a href="http://www.ee.surrey.ac.uk/Teaching/Unix/">Unix tutorial for beginners</a></em>. You must use <a href="https://scilifelab.github.io/courses/ngsintro/common/emu/">https://scilifelab.github.io/courses/ngsintro/common/emu/</a> <a href="https://bellard.org/jslinux/vm.html?url=https://bellard.org/jslinux/buildroot-x86.cfg">(or this mirror)</a> to try the commands in the tutorial, so that you don’t mess up any real world system. If you any questions regarding this tutorial contact: <em><a href="mailto:martin.dahlo@scilifelab.uu.se" class="email">martin.dahlo@scilifelab.uu.se</a></em>.</p>
</div>
<p>Make sure that you can read and write in the workshop folder. Go to <code>/proj/g2020006/nobackup/</code> and create a directory with your username. For example <code>mkdir jody</code>. You will work inside this directory for the workshop. For example <code>/proj/g2020006/nobackup/jody</code>. If you cannot write to the folder, the most likely reason is that you have not requested access to the workshop project via SUPR. This is described in step 1.</p>
<div class="boxy boxy-exclamation">
<p>Note that it may take an hour or so from request approval, before you can actually write to the folder. We will check before the workshop that all students have logged in and done this, so do not forget!</p>
</div>
</div>
<div id="tools" class="section level3">
<h3>Tools</h3>
<p><i class="fas fa-chevron-circle-right"></i> <strong>Download IGV</strong> (Integrated Genome Browser) from the <a href="https://www.broadinstitute.org/igv/">Broad Institute</a> on your own computer and make sure that you have human genome available. The will work with human data and visualise the results in the IGV genome browser.</p>
<p><a href="https://cloud.r-project.org/"><img src="data/common/r.png" /></a></p>
<p><i class="fas fa-chevron-circle-right"></i> <strong>Install R statistical software</strong> from <a href="https://cloud.r-project.org/">r-project.org</a>. If you have an old version of R and you do not use it, uninstall it and then install a newer version. Make sure you have one of the recent versions of R. One version older than the latest version is recommended. This is because all packages may not be updated for the very latest version of R. For Windows users, it is recommended that you DO NOT install to <code>C:\Program Files\R\</code>. Instead, install to <code>C:\R\</code>.</p>
<p><a href="https://www.rstudio.com/products/rstudio/download/"><img src="data/common/rstudio.png" /></a></p>
<p><i class="fas fa-chevron-circle-right"></i> <strong>Install <a href="https://www.rstudio.com/products/rstudio/download/">RStudio</a></strong>. RStudio provides you with tools like code editor with highlighting, project management, version control, package building, debugger, profiler and more.</p>
<p><i class="fas fa-chevron-circle-right"></i> <strong>Extra R packages</strong> used in the workshop exercises (if any) are listed below. It is recommended that you install this in advance.</p>
<p>For Linux and Mac users, you may have to install some extra OS specific libraries before installing the R packages. Pick the packages as needed based on your system below.</p>
<blockquote>
<p>libcurl
* deb: libcurl4-openssl-dev (Debian, Ubuntu, etc)
* rpm: libcurl-devel (Fedora, CentOS, RHEL)
* csw: libcurl_dev (Solaris)</p>
<p>libssl
* deb: libssl-dev (Debian, Ubuntu, etc)
* rpm: openssl-devel (Fedora, CentOS, RHEL)
* csw: libssl_dev (Solaris)
* brew: <a href="mailto:openssl@1.1" class="email">openssl@1.1</a> (Mac OSX)</p>
<p>libxml2
* deb: libxml2-dev (Debian, Ubuntu, etc)
* rpm: libxml2-devel (Fedora, CentOS, RHEL)
* csw: libxml2_dev (Solaris)</p>
</blockquote>
<p>Install the R packages. Simply copy and paste the code into R.</p>
<pre class="r"><code># install from cran
install.packages(c(&#39;BiocManager&#39;,&#39;dplyr&#39;,&#39;enrichR&#39;,&#39;ggplot2&#39;,&#39;ggpubr&#39;,&#39;pheatmap&#39;,&#39;pvclust&#39;,&#39;rafalib&#39;,&#39;stringr&#39;,&#39;tidyr&#39;))
# install from bioconductor
BiocManager::install(c(&#39;biomaRt&#39;,&#39;DESeq2&#39;,&#39;edgeR&#39;,&#39;fgsea&#39;))</code></pre>
<hr />
</div>
</div>
<div id="syllabus" class="section level2">
<h2>Syllabus</h2>
<p>The syllabus for this workshop are as follows.</p>
<ul>
<li>Introduction to RNA-Seq technology and workflow</li>
<li>Assessing quality of raw Illumina short reads using FastQC</li>
<li>Mapping short reads to reference genome using STAR aligner</li>
<li>Post alignment QC</li>
<li>Quantification of mapped reads using featureCounts</li>
<li>Exploratory data analyses using PCA and cluster analysis</li>
<li>Differential gene expression using DESeq2</li>
</ul>
<hr />
</div>
<div id="learning-outcomes" class="section level2">
<h2>Learning outcomes</h2>
<p>[To be added]</p>
<hr />
</div>
</div>
<div id="section" class="section level1">
<h1></h1>
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">
  
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p class="small" style="color:#bdbdbd;">
<b>2020</b> • NBIS • SciLifeLab
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
