<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Johan Reimegård • 23-Oct-2019" />


<title>Quality control</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><b>NBIS • Workshop on RNA-Seq</b></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="schedule.html">
    <span class="fa fa-clipboard-list"></span>
     
    Schedule
  </a>
</li>
<li>
  <a href="lab.html">
    <span class="fa fa-flask"></span>
     
    Lab
  </a>
</li>
<li>
  <a href="precourse.html">
    <span class="fa fa-pencil-alt"></span>
     
    Precourse
  </a>
</li>
<li>
  <a href="info.html">
    <span class="fa fa-info"></span>
     
    Info
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Quality control</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>
<h4 class="author"><b>Johan Reimegård</b> • 23-Oct-2019</h4>

</div>


<!-- rmd lab header -->
<!-- custom fonts -->
<p><link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;subset=latin-ext" rel="stylesheet"></p>
<p><br></p>
<p>In this tutorial we will go through some of the key steps in performing a quality control on your samples. We will start with the read based quality control, using FastQC, and continue with mapping based QC using RseqQC.</p>
<p>All the data you need for this lab is available in the folder: <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/QC/data/</code>.</p>
<p>This folder contains:</p>
<ul>
<li>Two FASTQ files, with mate pair libraries for sample 1 that is used in several of the other exercises, See section <strong>Lab</strong>.</li>
<li>BAM file and BAM-file index (<code>bam.bai</code>) for that sample mapped to the human genome using STAR.</li>
<li>count_table.txt - a table with number of reads per gene, using Ensembl annotations, created with HTseq-count.</li>
</ul>
<div id="before-mapping" class="section level1">
<h1><span class="header-section-number">1</span> Before mapping</h1>
<div id="fastqc-on-single-file" class="section level2">
<h2><span class="header-section-number">1.1</span> FastQC on single file</h2>
<p>FastQC aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to get a quick impression of whether your data has any problems of which you should be aware before doing any further analysis.</p>
<p>The main functions of FastQC are:</p>
<ul>
<li>Import of data from BAM, SAM or FASTQ files (any variant)</li>
<li>Providing a quick overview to tell you in which areas there may be problems</li>
<li>Summary graphs and tables to quickly assess your data</li>
<li>Export of results to an HTML-based permanent report</li>
<li>Offline operation to allow automated generation of reports without running the interactive application</li>
</ul>
<p>You can read more about the program and have a look at example reports at <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">the FastQC website</a>.</p>
<p><i class="fas fa-lightbulb"></i> This program can be used for any type of NGS data, not only RNA-seq.</p>
<p>To run FastQC on UPPMAX you first need to load the module:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="ex">module</span> load bioinfo-tools
<span class="ex">module</span> load FastQC/0.11.5

<span class="co"># To see help information on the FastQC package:</span>
<span class="ex">fastqc</span> --help
<span class="co"># Run for one FASTQ file:</span>
<span class="ex">fastqc</span> -o outdir fastqfile

<span class="co"># Run on multiple FASTQ files:</span>
<span class="ex">fastqc</span> -o outdir fastqfile1 fastqfile2 etc.

<span class="co"># You can use wildcards to run on all FASTQ files in a directory:</span>
<span class="ex">fastqc</span> -o outdir /proj/b2013006/INBOX/FASTQ/*fastq</code></pre></div>
</div>
<p>In this case, only run FastQC on one file and take a look at the output. We have already prepared the outputs for all of the other samples. These can be copied from UPPMAX at: <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/QC/fastQC/</code>.</p>
<p><i class="fas fa-clipboard-list"></i> Take a look at some other file and see if it look similar in quality.</p>
</div>
<div id="multiqc-on-fastqc-reports" class="section level2">
<h2><span class="header-section-number">1.2</span> MultiQC on FastQC reports</h2>
<p>MultiQC is a program that creates summaries over all samples for several different types of QC-measures. You can read more about the program <a href="http://multiqc.info/">here</a>. It will automatically look for output files from the supported tools and make a summary of them. You can either go to the folder where you have the Fastqc output or run it with the path to the folder.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="ex">module</span> load bioinfo-tools
<span class="ex">module</span> load MultiQC/1.5
<span class="ex">multiqc</span> /folder/with/FastQC_results/
<span class="co"># in this case the folder is /proj/uppstore2017171/courses/RNAseqWorkshop/downloads/QC/fastQC/</span></code></pre></div>
</div>
<p>This should create 1 folder named <code>multiqc_data</code> with some general stats, links to all files etc. And one file <code>multiqc_report.html</code>.</p>
<p><i class="fas fa-clipboard-list"></i> Have a look at the report you created. Alternatively, you can find the version we ran on UPPMAX at <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/QC/fastQC/multiqc_report.html</code>.</p>
</div>
</div>
<div id="after-mapping" class="section level1">
<h1><span class="header-section-number">2</span> After mapping</h1>
<p>These steps are to be carried out after the fastq files have been aligned to the reference.</p>
<div id="map-logs" class="section level2">
<h2><span class="header-section-number">2.1</span> Map logs</h2>
<p>The first step after you have finished your mapping is to get a general feel of how the mapping went. Most mapping programs produce some sort of summary output, either to a file or to standard out. For example, if using the mapper Bowtie you need to pipe that output to a file to see the summary statistics. In this case, the samples were mapped with STAR, that by default creates a file called <strong>Log.final.out</strong> in the mapping directory. Here is one example of <strong>Log.final.out</strong> content:</p>
<pre><code>                           Started job on |       Oct 16 20:21:39
                       Started mapping on |       Oct 16 20:27:04
                              Finished on |       Oct 16 20:29:14
 Mapping speed, Million of reads per hour |       366.35

                    Number of input reads |       13229276
                Average input read length |       202
                              UNIQUE READS:
             Uniquely mapped reads number |       11913568
                  Uniquely mapped reads % |       90.05%
                    Average mapped length |       198.41
                 Number of splices: Total |       9523918
      Number of splices: Annotated (sjdb) |       9443434
                 Number of splices: GT/AG |       9432792
                 Number of splices: GC/AG |       71488
                 Number of splices: AT/AC |       10675
         Number of splices: Non-canonical |       8963
                Mismatch rate per base, % |       0.33%
                   Deletion rate per base |       0.01%
                  Deletion average length |       1.75
                  Insertion rate per base |       0.01%
                 Insertion average length |       1.39
                       MULTI-MAPPING READS:
  Number of reads mapped to multiple loci |       356839
       % of reads mapped to multiple loci |       2.70%
  Number of reads mapped to too many loci |       2102
       % of reads mapped to too many loci |       0.02%
                            UNMAPPED READS:
 % of reads unmapped: too many mismatches |       0.00%
           % of reads unmapped: too short |       7.21%
               % of reads unmapped: other |       0.02%</code></pre>
<p>The most important parts to look at are the proportion of uniquely mapping, multi-mapping and unmapped reads. We ideally want the uniquely mapping reads to be as high as possible. Multi-mapping or unmapped reads could indicate poor quality of the reads, adapter contamination or other reasons for low quality scores.</p>
<p>Another key point is the mismatch and indel rates. If they are very high, this could indicate that there has been some problems during the sequencing or during the library preparation.</p>
</div>
<div id="multiqc-summary-of-logs" class="section level2">
<h2><span class="header-section-number">2.2</span> MultiQC summary of logs</h2>
<p>After mapping with star of all samples, we ran MultiQC to summarize all the logfiles. In this case we had a folder structure with <strong>sample_name/Log.final.out</strong>, so to make sure that MultiQC understands what is the sample name, we used the <code>-dd 2</code> command (e.g. it splits up the path and names the samples after the second last part).</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="co"># OBS! do not run now, just for reference</span>
<span class="ex">module</span> load bioinfo-tools
<span class="ex">module</span> load MultiQC/0.8
<span class="ex">multiqc</span> -d -dd 2 .</code></pre></div>
</div>
<p>You can find the output from that MultiQC report on UPPMAX <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/QC/output/multiqc_report_star.html</code>.</p>
</div>
<div id="rseqc" class="section level2">
<h2><span class="header-section-number">2.3</span> RSeQC</h2>
<p>The RSeQC package is one of many tools to get basic mapping statistics from your BAM files. This package provides a number of useful modules that can comprehensively evaluate high throughput sequence data, especially RNA-seq data. Some basic modules quickly inspect sequence quality, nucleotide composition bias, PCR bias and GC bias, while RNA-seq specific modules evaluate sequencing saturation, mapped reads distribution, coverage uniformity, strand specificity, etc. You can read more about the package at <a href="http://rseqc.sourceforge.net/">the RSeQC website</a>.</p>
<p>The RSeQC package contains many steps that are equivalent to FastQC analysis, e.g. read quality, sequence composition (NVC), GC-bias etc, but the results may be different since many of the low quality reads may not map to the genome and therefore will not be included in the BAM file.</p>
<p>Running all the QC steps takes a long time, so to save time, we only run the QC on a random selection of 10% of the reads. Random selection of reads can be performed with many different programs. Here we will use samtools:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="ex">samtools</span> view -b -s 0.1 Aligned.out.sorted.bam <span class="op">&gt;</span> Aligned.out.0.1.bam
<span class="co"># then index the bamfile</span>
<span class="co"># (it is already sorted since you extracted reads from a sorted BAM file)</span>
<span class="ex">samtools</span> index Aligned.out.0.1.bam</code></pre></div>
</div>
<p>The RSeQC package is already installed on UPPMAX. Load the package:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="ex">module</span> add bioinfo-tools
<span class="ex">module</span> add rseqc/2.6.4</code></pre></div>
</div>
<p>Some steps of the RSeQC package require a file with gene annotations in BED format. These can be downloaded from sources such as UCSC, RefSeq and Ensembl. In this case, the RSeQC team have already created annotation files in some common formats that can be downloaded from their website, but if you have data for a less studied organism you may need to create a BED-file on your own.</p>
<p>Two annotation files have already been downloaded into <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/QC/annotation</code> for you to use. These are: <strong>hg19.HouseKeepingGenes.bed</strong> and <strong>hg19_RefSeq.bed</strong>. The folder also contains a reduced annotation file <strong>hg19_RefSeq_top1000.bed</strong> to speed things up.</p>
<p>In this tutorial, we will not run all the different parts of the RSeQC package, only the most relevant ones for this experiment. The different scripts in the RSeQC package are well described on <a href="http://rseqc.sourceforge.net/">their website</a>, so read the instructions there and specify the input/output files to fit your file names and folder structure.</p>
<p>The steps that we are going to run are:</p>
<ol style="list-style-type: decimal">
<li>geneBody_coverage.py</li>
<li>inner_distance.py</li>
<li>junction_saturation.py</li>
<li>read_distribution.py</li>
</ol>
<p><i class="fas fa-exclamation-circle"></i> The <strong>geneBody_coverage.py</strong> script takes a very long time to run, so we have created a subsection of annotations to run it on. Use the file <code>hg19_RefSeq_top1000.bed</code>. This file was created with the command:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">head</span> -n 1000 hg19_RefSeq.bed <span class="op">&gt;</span> hg19_RefSeq_top1000.bed</code></pre></div>
</div>
<p><i class="fas fa-lightbulb"></i> When running <strong>read_distribution.py</strong>, an outfile cannot be specified. Instead you need to pipe (<code>&gt;</code>) the output to a file, or look at the output in the terminal.</p>
<p><i class="fas fa-clipboard-list"></i> Run RSeQC for one sample and have a look at your output.</p>
<ul>
<li>Do most of your reads map to genes?</li>
<li>Do you have even coverage along the genes?</li>
<li>Do the reads cover most splice junctions?</li>
<li>Based on the inner distance plots, what do you think the average fragment size of the libraries was?</li>
</ul>
</div>
<div id="multiqc-summary-of-rseqc-output" class="section level2">
<h2><span class="header-section-number">2.4</span> MultiQC summary of RSeQC output</h2>
<p>We have ran RSeQC and MultiQC on all the samples in the project. The summary report from MultiQC can be found on UPPMAX <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/QC/output/multiqc_report_rseqc.html</code> .</p>
<p>It was created using commands:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="ex">multiqc</span> -f -d -dd 3 .
<span class="co"># since folder structure like: sample_name/qc/read_distribution.txt and so on for the other file types.</span></code></pre></div>
</div>
<p><i class="fas fa-clipboard-list"></i> Have a look at the reports. What is your conclusion, do your samples look good? Is there anything that looks strange in any sample, or do you feel comfortable using all the samples in your analysis?</p>
<p><strong>End of document</strong></p>
</div>
</div>


<div class="footer">
<div class="container" style="padding-top:5px;padding-bottom:5px;">
<p>
<span style="float:left; vertical-align:middle">
<b>2019 </b><a href="https://nbis.se/">NBIS</a> • <a href="https://www.scilifelab.se/">SciLifeLab</a>
</span>
<span style="float:right; vertical-align:middle">
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://nbis.se/"><i class="fas fa-globe-americas"></i></a>
</span>
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://twitter.com/NBISwe"><i class="fab fa-twitter"></i></a>
</span>
<span class="footericon" style="padding-left:4px">
<a href="https://www.linkedin.com/company/nbisweden/"><i class="fab fa-linkedin"></i></a>
</span>
</span>
</p>
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
