<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Gene Function Annotation</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-6.4.2/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">






</div>


<!-- rmd lab header -->
<p><br></p>
<div id="loading-data" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Loading data</h1>
<div class="boxy boxy-exclamation boxy-yellow">
<p>Set <code>~/RNAseq/labs/</code> as your working directory (i.e.Â your working directory should be the directory where you saved this .Rmd file). Create a subdirectory named <code>data</code> in this directory (<code>~/RNAseq/labs/data/</code>) for input and output files.</p>
</div>
<p>Loading packages and data.</p>
<pre class="r"><code>library(pheatmap)
library(rafalib)
library(DESeq2)
library(pvclust)
library(biomaRt)
library(enrichR)
library(fgsea)

# source download function
source(&quot;https://raw.githubusercontent.com/NBISweden/workshop-RNAseq/master/assets/scripts.R&quot;)


# read data
download_data(&quot;data/gene_counts.csv&quot;)
data &lt;- read.csv(&quot;data/gene_counts.csv&quot;,row.names = 1)
dim(data)

# read metadata
download_data(&quot;data/metadata_raw.csv&quot;)
metadata &lt;- read.csv(&quot;data/metadata_raw.csv&quot;,row.names = 1,stringsAsFactors=T)</code></pre>
<pre><code>## [1] 10550     6</code></pre>
<p>Here you would select genes based on p-values and logFC you obtained from differential expression.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code>res &lt;- read.csv(&quot;data/dge_results.csv&quot;,row.names=1)
res1 &lt;- na.omit(res)
selected_genes &lt;- rownames(res1)[(res1$pvalue &lt; 0.001) &amp; (abs(res1$log2FoldChange) &gt; 1)]</code></pre>
</div>
<div class="boxy boxy-lightbulb">
<p>There are many ways of selecting gene lists. One could for example select only the top 500 variable genes. CPM counts are converted to log scale and row-wise variance is computed.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code># Gene selection
vars &lt;- apply(data,1,var)
vars &lt;- sort(vars,decreasing=T)
top_var &lt;- names(vars)[1:500]</code></pre>
</div>
</div>
<div id="clustering-on-heatmap" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Clustering on Heatmap</h2>
<p>Now that you understand the concepts of hierarchical clustering both at the sample and at the gene level, we can use a heatmap function to explore the visual consequences of clustering. Here, we can make use of the <code>pheatmap()</code> function, which by default will do the clustering of the rows and columns.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code>cl &lt;- pheatmap( data[rownames(data) %in% selected_genes,],scale=&quot;row&quot;,color=colorRampPalette(c(&quot;navy&quot;,&quot;white&quot;,&quot;firebrick&quot;))(90),border_color=NA,cluster_cols=F,cutree_rows=2)

gene_clusters &lt;- cutree(cl$tree_row,k=2)</code></pre>
<img src="lab_functional_files/figure-html/unnamed-chunk-5-1.png" width="480" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
</div>
<div id="functional-analysis" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Functional Analysis</h1>
<table>
<thead>
<tr class="header">
<th>R-packages</th>
<th>Comments</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>topGO</td>
<td>GO</td>
</tr>
<tr class="even">
<td>goana</td>
<td>GO</td>
</tr>
<tr class="odd">
<td>GOseq</td>
<td>GO</td>
</tr>
<tr class="even">
<td>topKEGG</td>
<td>KEGG</td>
</tr>
<tr class="odd">
<td>kegga</td>
<td>KEGG</td>
</tr>
<tr class="even">
<td>enrichR</td>
<td>GO, KEGG, many others</td>
</tr>
<tr class="odd">
<td>piano</td>
<td>GO, KEGG, GSEA, many others, enrichment consensus</td>
</tr>
<tr class="even">
<td>ClusterProfiler</td>
<td>GO, KEGG, GSEA, many others, nice plots!</td>
</tr>
<tr class="odd">
<td>Pathview</td>
<td>Nice visualization for KEGG pathways</td>
</tr>
<tr class="even">
<td>fgsea</td>
<td>GSEA</td>
</tr>
</tbody>
</table>
<div id="enrichr" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> enrichR</h2>
<p>You can list all available databases by using the command <code>listEnrichrDbs()</code> function.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code>genes_cluster1 &lt;- names(gene_clusters)[gene_clusters == 1]
genes_cluster2 &lt;- names(gene_clusters)[gene_clusters == 2]

head(data[genes_cluster1, ])
head(data[genes_cluster2, ])</code></pre>
</div>
<div id="go-enrichment" class="section level3" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> GO enrichment</h3>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code>go_cluster &lt;- enrichr(genes = genes_cluster2,databases = &quot;GO_Biological_Process_2018&quot;)
go_cluster &lt;- go_cluster$GO_Biological_Process_2018
go_cluster &lt;- go_cluster[order(go_cluster$P.value),]
go_cluster[1:5,]

{
  mypar(1,1,mar=c(2,25,2,2))
  barplot(-log10(go_cluster$P.value[15:1]),horiz=T,border=F,yaxs=&quot;i&quot;,
           names.arg=go_cluster$Term[15:1],las=1,cex.names=0.8)
  abline(v=0,lwd=2)
}</code></pre>
<img src="lab_functional_files/figure-html/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto auto auto 0;" />
</div>
<p><i class="fas fa-comments"></i> <strong>Exploratory Questions</strong></p>
<ul>
<li>Which is the most enriched GO term for cluster1?</li>
<li>How many genes from your data set were detected in this most GO term?
What is the percentage of genes out of the total list that defines the GO term?</li>
<li>Which genes from your data set belong to this most enriched GO term?</li>
<li>Some genes make part of several GO terms, can you list some?</li>
</ul>
</div>
<div id="kegg-enrichment" class="section level3" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> KEGG enrichment</h3>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code>kegg_cluster &lt;- enrichr(genes=genes_cluster2,databases=&quot;KEGG_2019_Human&quot;)
kegg_cluster &lt;- kegg_cluster$KEGG_2019_Human
kegg_cluster &lt;- kegg_cluster[order(kegg_cluster$P.value),]
kegg_cluster[1:5,]

{
  mypar(1,1,mar=c(2,25,2,2))
  barplot(-log10(kegg_cluster$P.value[15:1]),horiz=T,border=F,yaxs=&quot;i&quot;,
           names.arg=kegg_cluster$Term[15:1],las=1,cex.names=0.8)
  abline(v=0,lwd=2)
}</code></pre>
<img src="lab_functional_files/figure-html/unnamed-chunk-8-1.png" width="960" style="display: block; margin: auto auto auto 0;" />
</div>
<p><i class="fas fa-comments"></i> <strong>Exploratory Questions</strong></p>
<ul>
<li>Which is the most enriched KEGG pathway for cluster1?</li>
<li>How many genes from your data set were detected in this most enriched KEGG pathway ?</li>
<li>What is the percentage of genes out of the total list that defines the KEGG pathway ?</li>
<li>Which genes from your data set belong to this most enriched KEGG pathway ?</li>
<li>Some genes make part of several KEGG pathways, can you list some?</li>
<li>Taking the results from GO term enrichment and the KEGG enrichment, what is a most likely function happening in your data set?</li>
</ul>
</div>
</div>
<div id="gsea" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> GSEA</h2>
<p>MSigDB is one of the largest curated databases of signatures and pathways created by the BROAD INSTITUTE. The Molecular Signatures Database (MSigDB) is accessible <a href="http://software.broadinstitute.org/gsea/msigdb/index.jsp" target="_blank">here</a>.</p>
<p>For this exercise, you can download a couple of data sets:</p>
<table>
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead>
<tr class="header">
<th>Database</th>
<th>Link to Download</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>KEGG</td>
<td><a href="http://software.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/6.2/c2.cp.kegg.v6.2.symbols.gmt" target="_blank">c2.cp.kegg.v6.2.symbols.gmt</a></td>
</tr>
<tr class="even">
<td>GO_BP</td>
<td><a href="http://software.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/6.2/c5.bp.v6.2.symbols.gmt" target="_blank">c5.bp.v6.2.symbols.gmt</a></td>
</tr>
<tr class="odd">
<td>HALLMARK</td>
<td><a href="http://software.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/6.2/h.all.v6.2.symbols.gmt" target="_blank">h.all.v6.2.symbols.gmt</a></td>
</tr>
</tbody>
</table>
<p>As you could already notice, the differences in gene expression between days 0 to 7 are very clear. Here we will illustrate a gene set enrichment to explore which pathways are UP or DOWN regulated.</p>
<div class="boxy boxy-exclamation boxy-yellow">
<p>Here you would use the logFC values you obtained from differential expression. But for simplicity and reduce dependency on other parts of the course, we are using a simple difference in mean expression.</p>
</div>
<pre class="r"><code># if folder and files doesn&#39;t exist, download it
if(!dir.exists(&quot;data/MSigDB_files/&quot;)) {
  dir.create(&quot;data/MSigDB_files&quot;)
  download.file(url=&quot;https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c2.cp.kegg.v6.2.symbols.gmt.txt&quot;,&quot;./data/MSigDB_files/c2.cp.kegg.v6.2.symbols.gmt.txt&quot;)
  download.file(url=&quot;https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c3.tft.v6.2.symbols.gmt.txt&quot;,&quot;./data/MSigDB_files/c3.tft.v6.2.symbols.gmt.txt&quot;)
  download.file(url=&quot;https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/h.all.v6.2.symbols.gmt.txt&quot;,&quot;./data/MSigDB_files/h.all.v6.2.symbols.gmt.txt&quot;)
  download.file(url=&quot;https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c2.cp.v6.2.symbols.gmt.txt&quot;,&quot;./data/MSigDB_files/c2.cp.v6.2.symbols.gmt.txt&quot;)
  download.file(url=&quot;https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c5.bp.v6.2.symbols.gmt.txt&quot;,&quot;./data/MSigDB_files/c5.bp.v6.2.symbols.gmt.txt&quot;)
}</code></pre>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code># Create a gene rank based on the gene expression
gene_rank &lt;- setNames( res$log2FoldChange,casefold(rownames(res),upper=T) )

# Load hallmark pathways
hallmark_pathways &lt;- gmtPathways(&quot;data/MSigDB_files/h.all.v6.2.symbols.gmt.txt&quot;)</code></pre>
</div>
<p>Once our list of genes are sorted (from highest to lowest expression in day 7), we can proceed with the enrichment itself. Here, we will be using the <code>fgsea()</code> function. This will result in a table containing information for several pathways. We can then sort and filter those pathways to visualize only the top ones. You can select/filter them by either <code>p-value</code> or normalized enrichemnet score (<code>NES</code>).</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code># Perform enrichemnt analysis
fgseaRes &lt;- fgsea( pathways=hallmark_pathways, stats=gene_rank, minSize=15, maxSize=500, nperm=10000)

# Filter the results table to show only the top 10 UP and DOWN regulated processes (optional)
topPathwaysUp &lt;- fgseaRes[ES &gt; 0][head(order(pval), n=10), pathway]
topPathwaysDown &lt;- fgseaRes[ES &lt; 0][head(order(pval), n=10), pathway]
topPathways &lt;- c(topPathwaysUp, rev(topPathwaysDown))

# Nice summary table (shown as a plot)
plotGseaTable(hallmark_pathways[topPathways], gene_rank, fgseaRes, gseaParam = 0.5)</code></pre>
<img src="lab_functional_files/figure-html/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto auto auto 0;" />
</div>
<p><i class="fas fa-comments"></i> <strong>Exploratory Questions</strong></p>
<ul>
<li>Which is the most significantly up-regulated pathway at day 7?</li>
<li>What do the length of the black bars mean?</li>
<li>Is there any pathway that is significantly down-regulated at 0? Which columns tells you what is Up of downregulated?</li>
<li>Which is the most enriched pathway at day 0?</li>
<li>What happens if we use <code>-res$log2FoldChange</code> as gene rank ? What changes and why?</li>
</ul>
<p>Checking for individual enrichment:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code>topPathways
source(&quot;https://raw.githubusercontent.com/czarnewski/niceRplots/master/R/helper_functions.R&quot;)
mypar(2,3)

# Nice plot to see enrichment of one specific pathway
p &lt;- list()
for (i in c(topPathwaysUp[1:3],topPathwaysDown[3:1])){
  plot_enrich(i,hallmark_pathways,gene_rank)
  # p[[i]] &lt;- plotEnrichment(hallmark_pathways[[i]],  gene_rank) + ggplot2::ggtitle(i)
}
# p</code></pre>
<img src="lab_functional_files/figure-html/unnamed-chunk-12-1.png" width="960" style="display: block; margin: auto auto auto 0;" />
</div>
<p><i class="fas fa-comments"></i> <strong>Exploratory Questions</strong></p>
<ul>
<li>What do the black lines in the X-axis mean?</li>
<li>By looking at the green line distribution, and the comparison made (day00 - day07), at which time point is this pathways enriched.</li>
</ul>
<p><i class="fas fa-comments"></i> <strong>Exploratory Questions (optional)</strong></p>
<ul>
<li>What if instead of the hallmark gene set, you use the GO? Try that.</li>
<li>How different are the results from EnrichR and GSEA using GO annotation? Remember to use the same comparison (i.e.Â day07 vs day00).</li>
</ul>
</div>
<div id="online-enrichment-tools" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Online enrichment tools:</h2>
<p>Below you can find a list with the most commonly used online enrichment tools. Each one differs a bit on what they can do. The packages are sorted in order, we would like you to work on the workshop.</p>
<table>
<colgroup>
<col width="25%" />
<col width="25%" />
<col width="25%" />
<col width="25%" />
</colgroup>
<thead>
<tr class="header">
<th>Package</th>
<th>link</th>
<th>Database</th>
<th>Comment</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Enrichr</td>
<td><a href="http://amp.pharm.mssm.edu/Enrichr/" class="uri">http://amp.pharm.mssm.edu/Enrichr/</a></td>
<td>GO, KEGG, TF, many others</td>
<td>Extensive libraries</td>
</tr>
<tr class="even">
<td>GOrilla</td>
<td><a href="http://cbl-gorilla.cs.technion.ac.il" class="uri">http://cbl-gorilla.cs.technion.ac.il</a></td>
<td>GO</td>
<td>Support for REVIGO</td>
</tr>
<tr class="odd">
<td>REVIGO</td>
<td><a href="http://revigo.irb.hr" class="uri">http://revigo.irb.hr</a></td>
<td>GO</td>
<td>Summarises redundancy</td>
</tr>
<tr class="even">
<td>DAVID</td>
<td><a href="https://david.ncifcrf.gov" class="uri">https://david.ncifcrf.gov</a></td>
<td>GO, KEGG, TF, many others</td>
<td>*Not updated</td>
</tr>
<tr class="odd">
<td>KEGG</td>
<td><a href="https://www.genome.jp/kegg/" class="uri">https://www.genome.jp/kegg/</a></td>
<td>KEGG</td>
<td>Shows the pathways</td>
</tr>
<tr class="even">
<td>Reactome</td>
<td><a href="https://www.reactome.org" class="uri">https://www.reactome.org</a></td>
<td>KEGG-like</td>
<td>Shows the pathways/reactions</td>
</tr>
<tr class="odd">
<td>Panther</td>
<td><a href="http://www.pantherdb.org/about.jsp" class="uri">http://www.pantherdb.org/about.jsp</a></td>
<td>GO</td>
<td>Evolutionary conserved GO annotation</td>
</tr>
</tbody>
</table>
<p>In case you want to test the online tools above, you can use the code below to copy the gene vector into memory, and then paste it to the webtools. Should work for both Mac and Windows users.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<pre class="r"><code>clip &lt;- pipe(&quot;pbcopy&quot;,&quot;w&quot;)
write.table(genes_cluster1, file=clip, sep = &#39;\t&#39;, row.names = FALSE)         
close(clip)</code></pre>
</div>
</div>
<div id="protein-protein-interactions" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Protein-protein interactions</h2>
<p>In the course we will not have much time to work on gene expression networks. However, there are some nice databases that allow you to easily visualize protein-protein interactions. Please paste the list of genes using the command above into those sites to have a visual insight on how the Up- or Down- regulated genes interact with each other.</p>
<table>
<thead>
<tr class="header">
<th>Database</th>
<th>Link</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GeneMANIA</td>
<td><a href="https://genemania.org" class="uri">https://genemania.org</a></td>
</tr>
<tr class="even">
<td>STRING-DB</td>
<td><a href="https://string-db.org" class="uri">https://string-db.org</a></td>
</tr>
<tr class="odd">
<td>MIST</td>
<td><a href="https://fgrtools.hms.harvard.edu/MIST/help.jsp" class="uri">https://fgrtools.hms.harvard.edu/MIST/help.jsp</a></td>
</tr>
</tbody>
</table>
<hr />
</div>
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">

<div class="footer-left col-sm-6">
<p class="small">
<b>2024</b> â¢ <a href="https://nbis.se/" target="_blank">NBIS</a> â¢ <a href="https://www.scilifelab.se/" target="_blank">SciLifeLab</a>
</p>
</div>

<div class="footer-right col-sm-6">
<p>
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/" target="_blank">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe" target="_blank"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/" target="_blank">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
