<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Intro to R</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:22px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Intro to R</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>

</div>


<!-- rmd lab header -->
<p><br></p>
<p>R is a programming language for statistical computing, and data wrangling. It is open-source, widely used in data science, has a wide range of functions and algorithms for graphing and data analyses.</p>
<div id="assignment-operator" class="section level1">
<h1><span class="header-section-number">1</span> Assignment operator</h1>
<p>Variables are assigned usually using the <code>&lt;-</code> operator. The <code>=</code> operator also works in a similar way for most part.</p>
<pre class="r"><code>x &lt;- 4
x = 4
x</code></pre>
<pre><code>## [1] 4</code></pre>
</div>
<div id="arithmetic-operators" class="section level1">
<h1><span class="header-section-number">2</span> Arithmetic operators</h1>
<p>The commonly used arithmetic operators are shown below returning a number.</p>
<pre class="r"><code>x &lt;- 4
y &lt;- 2

# add
x + y

# subtract
x - y

# multiply
x * y

# divide
x / y

# modulus
x %% y

# power
x ^ y</code></pre>
<pre><code>## [1] 6
## [1] 2
## [1] 8
## [1] 2
## [1] 0
## [1] 16</code></pre>
</div>
<div id="logical-operators" class="section level1">
<h1><span class="header-section-number">3</span> Logical operators</h1>
<p>Logical operators return a logical <code>TRUE</code> or <code>FALSE</code>.</p>
<pre class="r"><code># equal to?
x == y

# not equal to?
x != y

# greater than?
x &gt; y

# less than?
x &lt; y

# greater than or equal to?
x &gt;= y

# less than or equal to?
x &lt;= y</code></pre>
<pre><code>## [1] FALSE
## [1] TRUE
## [1] TRUE
## [1] FALSE
## [1] TRUE
## [1] FALSE</code></pre>
</div>
<div id="data-types" class="section level1">
<h1><span class="header-section-number">4</span> Data types</h1>
<pre class="r"><code>class(1)
class(&quot;hello&quot;)
class(T)</code></pre>
<pre><code>## [1] &quot;numeric&quot;
## [1] &quot;character&quot;
## [1] &quot;logical&quot;</code></pre>
<pre class="r"><code>x &lt;- c(2,3,4,5,6)
y &lt;- c(&quot;a&quot;,&quot;c&quot;,&quot;d&quot;,&quot;e&quot;)
z &lt;- factor(c(&quot;a&quot;,&quot;c&quot;,&quot;d&quot;,&quot;e&quot;))
class(z)</code></pre>
<pre><code>## [1] &quot;factor&quot;</code></pre>
<pre class="r"><code>x &lt;- matrix(c(2,3,4,5,6,7),nrow=3,ncol=2)
class(x)
str(x)</code></pre>
<pre><code>## [1] &quot;matrix&quot; &quot;array&quot; 
##  num [1:3, 1:2] 2 3 4 5 6 7</code></pre>
<pre class="r"><code>dfr &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
print(dfr)</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c</code></pre>
<pre class="r"><code>class(dfr)</code></pre>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<pre class="r"><code>str(dfr)</code></pre>
<pre><code>## &#39;data.frame&#39;:    3 obs. of  2 variables:
##  $ x: int  1 2 3
##  $ y: chr  &quot;a&quot; &quot;b&quot; &quot;c&quot;</code></pre>
</div>
<div id="accessors" class="section level1">
<h1><span class="header-section-number">5</span> Accessors</h1>
<p>Vectors positions can be accessed using <code>[]</code>. R follows 1-based indexing.</p>
<pre class="r"><code>x &lt;- c(2,3,4,5,6)
x
x[2]</code></pre>
<pre><code>## [1] 2 3 4 5 6
## [1] 3</code></pre>
<p>Dataframe or matrix positions can be accessed using <code>[]</code> specifying row and column like <code>[row,column]</code>.</p>
<pre class="r"><code>dfr &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr
dfr[1,]
dfr[,1]
dfr[2,2]</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   x y
## 1 1 a
## [1] 1 2 3
## [1] &quot;b&quot;</code></pre>
</div>
<div id="functions" class="section level1">
<h1><span class="header-section-number">6</span> Functions</h1>
<pre class="r"><code># generate 10 random numbers between 1 and 200
x &lt;- sample(x=1:200,10)
x

# length
length(x)

# sum
sum(x)

# mean
mean(x)

# median
median(x)

# min
min(x)

# log
log(x)

# exponent
exp(x)

# square-root
sqrt(x)

# round
round(x)

# sort
sort(x)</code></pre>
<pre><code>##  [1] 123  10 132 148 168  48  81 158  34 196
## [1] 10
## [1] 1098
## [1] 109.8
## [1] 127.5
## [1] 10
##  [1] 4.812184 2.302585 4.882802 4.997212 5.123964 3.871201 4.394449 5.062595
##  [9] 3.526361 5.278115
##  [1] 2.619517e+53 2.202647e+04 2.122617e+57 1.886181e+64 9.151093e+72
##  [6] 7.016736e+20 1.506097e+35 4.154590e+68 5.834617e+14 1.323483e+85
##  [1] 11.090537  3.162278 11.489125 12.165525 12.961481  6.928203  9.000000
##  [8] 12.569805  5.830952 14.000000
##  [1] 123  10 132 148 168  48  81 158  34 196
##  [1]  10  34  48  81 123 132 148 158 168 196</code></pre>
<p>Some useful string functions.</p>
<pre class="r"><code>a &lt;- &quot;sunny&quot;
b &lt;- &quot;day&quot;

# join
paste(a, b)

# find a pattern
grep(&quot;sun&quot;, a)

# number of characters
nchar(&quot;sunny&quot;)

# to uppercase
toupper(&quot;sunny&quot;)

# to lowercase
tolower(&quot;SUNNY&quot;)

# replace pattern
sub(&quot;sun&quot;, &quot;fun&quot;, &quot;sunny&quot;)

# substring
substr(&quot;sunny&quot;, start=1, stop=3)</code></pre>
<pre><code>## [1] &quot;sunny day&quot;
## [1] 1
## [1] 5
## [1] &quot;SUNNY&quot;
## [1] &quot;sunny&quot;
## [1] &quot;funny&quot;
## [1] &quot;sun&quot;</code></pre>
<p>Some general functions</p>
<pre class="r"><code>print(&quot;hello&quot;)
print(&quot;world&quot;)
cat(&quot;hello&quot;)
cat(&quot; world&quot;)
cat(&quot;\nhello\nworld&quot;)</code></pre>
<pre><code>## [1] &quot;hello&quot;
## [1] &quot;world&quot;
## hello world
## hello
## world</code></pre>
</div>
<div id="merging" class="section level1">
<h1><span class="header-section-number">7</span> Merging</h1>
<p>Two strings can be joined together using <code>paste()</code>.</p>
<pre class="r"><code>a &lt;- &quot;sunny&quot;
b &lt;- &quot;day&quot;

paste(a, b)
paste(a, b, sep=&quot;-&quot;)</code></pre>
<pre><code>## [1] &quot;sunny day&quot;
## [1] &quot;sunny-day&quot;</code></pre>
<p>The function <code>c()</code> is used to concatenate objects.</p>
<pre class="r"><code>a &lt;- &quot;sunny&quot;
b &lt;- &quot;day&quot;

c(a,b)</code></pre>
<pre><code>## [1] &quot;sunny&quot; &quot;day&quot;</code></pre>
<p>The function <code>cbind()</code> is used to join two dataframes column-wise.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr2 &lt;- data.frame(p = 4:6, q = c(&quot;d&quot;, &quot;e&quot;, &quot;f&quot;))
dfr1
dfr2

cbind(dfr1,dfr2)</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   p q
## 1 4 d
## 2 5 e
## 3 6 f
##   x y p q
## 1 1 a 4 d
## 2 2 b 5 e
## 3 3 c 6 f</code></pre>
<p>Similarily, <code>rbind()</code> is used to join two dataframes row-wise.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr2 &lt;- data.frame(x = 4:6, y = c(&quot;d&quot;, &quot;e&quot;, &quot;f&quot;))
dfr1
dfr2

rbind(dfr1,dfr2)</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   x y
## 1 4 d
## 2 5 e
## 3 6 f
##   x y
## 1 1 a
## 2 2 b
## 3 3 c
## 4 4 d
## 5 5 e
## 6 6 f</code></pre>
<p>Two dataframes can be merged based on a shared column using the <code>merge()</code> function.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:4, p = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;,&quot;d&quot;))
dfr2 &lt;- data.frame(x = 3:6, q = c(&quot;l&quot;, &quot;m&quot;, &quot;n&quot;,&quot;o&quot;))
dfr1
dfr2

merge(dfr1,dfr2,by=&quot;x&quot;)
merge(dfr1,dfr2,by=&quot;x&quot;,all.x=T)
merge(dfr1,dfr2,by=&quot;x&quot;,all.y=T)
merge(dfr1,dfr2,by=&quot;x&quot;,all=T)</code></pre>
<pre><code>##   x p
## 1 1 a
## 2 2 b
## 3 3 c
## 4 4 d
##   x q
## 1 3 l
## 2 4 m
## 3 5 n
## 4 6 o
##   x p q
## 1 3 c l
## 2 4 d m
##   x p    q
## 1 1 a &lt;NA&gt;
## 2 2 b &lt;NA&gt;
## 3 3 c    l
## 4 4 d    m
##   x    p q
## 1 3    c l
## 2 4    d m
## 3 5 &lt;NA&gt; n
## 4 6 &lt;NA&gt; o
##   x    p    q
## 1 1    a &lt;NA&gt;
## 2 2    b &lt;NA&gt;
## 3 3    c    l
## 4 4    d    m
## 5 5 &lt;NA&gt;    n
## 6 6 &lt;NA&gt;    o</code></pre>
</div>
<div id="packages" class="section level1">
<h1><span class="header-section-number">8</span> Packages</h1>
<p>R packages extend the functionality of base R. R packages are stored in repositories of which the most commonly used is called <a href="https://cran.r-project.org/">CRAN</a> (The Comprehensive R Archive Network).</p>
<p>Packages are installed using the function <code>install.packages()</code>. Let’s install the graphics and plotting package <code>ggplot2</code> which will be useful in later sections.</p>
<pre class="r"><code>install.packages(&quot;ggplot2&quot;,dependencies=TRUE)</code></pre>
<p>Packages on BioConductor can be installed as follows:</p>
<pre class="r"><code>source(&quot;https://bioconductor.org/biocLite.R&quot;)
biocLite(&quot;biomaRt&quot;)</code></pre>
<p>Packages on GitHub can be installed using the function <code>install_github()</code> from package <code>devtools</code>.</p>
<p>Packages can also be installed from a local zipped file by providing a local path ans setting <code>type="source"</code>.</p>
<pre class="r"><code>install.packages(&quot;./dir/package.zip&quot;,type=&quot;source&quot;)</code></pre>
<p>Inside RStudio, installing packages is much easier. Go to the <strong>Packages</strong> tab and click <strong>Install</strong>. In the window that opens up, you can find your package by typing into the <strong>Packages</strong> field and clicking <strong>Install</strong>. Bioconductor packages can be added to this list by setting it using <code>setRepositories()</code>.</p>
</div>
<div id="graphics" class="section level1">
<h1><span class="header-section-number">9</span> Graphics</h1>
<div id="base" class="section level2">
<h2><span class="header-section-number">9.1</span> Base</h2>
<p>R is an excellent tool for creating graphs and plots. The graphic capabilities and functions provided by the base R installation is called the base R graphics. Numerous packages exist to extend the functionality of base graphics.</p>
<p>We can try out plotting a few of the common plot types. Let’s start with a scatterplot. First we create a <code>data.frame</code> as this is the most commonly used data object.</p>
<pre class="r"><code>dfr &lt;- data.frame(a=sample(1:100,10),b=sample(1:100,10))</code></pre>
<p>Now we have a dataframe with two continuous variables that can be plotted against each other.</p>
<pre class="r"><code>plot(dfr$a,dfr$b)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-24-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>This is probably the simplest and most basic plots. We can modify the x and y axis labels.</p>
<pre class="r"><code>plot(dfr$a,dfr$b,xlab=&quot;Variable a&quot;,ylab=&quot;Variable b&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-25-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>We can change the point to a line.</p>
<pre class="r"><code>plot(dfr$a,dfr$b,xlab=&quot;Variable a&quot;,ylab=&quot;Variable b&quot;,type=&quot;b&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-26-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>Let’s add a categorical column to our dataframe.</p>
<pre class="r"><code>dfr$cat &lt;- rep(c(&quot;C1&quot;,&quot;C2&quot;),each=5)</code></pre>
<p>And then colour the points by category.</p>
<pre class="r"><code># subset data
dfr_c1 &lt;- subset(dfr,dfr$cat == &quot;C1&quot;)
dfr_c2 &lt;- subset(dfr,dfr$cat == &quot;C2&quot;)

plot(dfr_c1$a,dfr_c1$b,xlab=&quot;Variable a&quot;,ylab=&quot;Variable b&quot;,col=&quot;red&quot;,pch=1)
points(dfr_c2$a,dfr_c2$b,col=&quot;blue&quot;,pch=2)

legend(x=&quot;topright&quot;,legend=c(&quot;C1&quot;,&quot;C2&quot;),
       col=c(&quot;red&quot;,&quot;blue&quot;),pch=c(1,2))</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-28-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>Let’s create a barplot.</p>
<pre class="r"><code>ldr &lt;- data.frame(a=letters[1:10],b=sample(1:50,10))
barplot(ldr$b,names.arg=ldr$a)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-29-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
<div id="grid" class="section level2">
<h2><span class="header-section-number">9.2</span> Grid</h2>
<p>Grid graphics have a completely different underlying framework compared to base graphics. Generally, base graphics and grid graphics cannot be plotted together. The most popular grid-graphics based plotting library is <strong>ggplot2</strong>.</p>
<p>Let’s create the same plot are before using <strong>ggplot2</strong>. Make sure you have the package installed.</p>
<pre class="r"><code>library(ggplot2)

ggplot(dfr,aes(x=a,y=b,colour=cat))+
  geom_point()+
  labs(x=&quot;Variable a&quot;,y=&quot;Variable b&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-30-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>It is generally easier and more consistent to create plots using the ggplot2 package compared to the base graphics.</p>
<p>Let’s create a barplot as well.</p>
<pre class="r"><code>ggplot(ldr,aes(x=a,y=b))+
  geom_bar(stat=&quot;identity&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-31-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
</div>
<div id="inputoutput" class="section level1">
<h1><span class="header-section-number">10</span> Input/Output</h1>
<p>Input and output of data and images is an important aspect with data analysis.</p>
<div id="text" class="section level2">
<h2><span class="header-section-number">10.1</span> Text</h2>
<p>Data can come in a variety of formats which needs to be read into R and converted to an R data type.</p>
<p>Text files are the most commonly used input. Text files can be read in using the function <code>read.table</code>. We have a sample file to use: <strong>iris.txt</strong>.</p>
<pre class="r"><code>dfr &lt;- read.table(&quot;iris.txt&quot;,header=TRUE,stringsAsFactors=F)</code></pre>
<p>This reads in a tab-delimited text file with a header. The argument <code>sep='\t'</code> is set by default to specify that the delimiter is a tab. <code>stringsAsFactors=F</code> setting ensures that character columns are not automatically converted to factors.</p>
<p>It’s always a good idea to check the data after import.</p>
<pre class="r"><code>head(dfr)</code></pre>
<pre class="r"><code>str(dfr)</code></pre>
<p>Check <code>?read.table</code> for other wrapper functions to read in text files.</p>
<p>Let’s filter this data.frame and create a new dataset.</p>
<pre class="r"><code>dfr1 &lt;- dfr[dfr$Species == &quot;setosa&quot;,]</code></pre>
<p>And we can write this as a text file.</p>
<pre class="r"><code>write.table(dfr1,&quot;iris-setosa.txt&quot;,sep=&quot;\t&quot;,row.names=F,quote=F)</code></pre>
<p><code>sep="\t"</code> sets the delimiter to tab. <code>row.names=F</code> denotes that rownames should not be written. <code>quote=F</code> specifies that doubles must not be placed around strings.</p>
</div>
<div id="images" class="section level2">
<h2><span class="header-section-number">10.2</span> Images</h2>
<p>Let’s take a look at saving plots.</p>
</div>
<div id="base-graphics" class="section level2">
<h2><span class="header-section-number">10.3</span> Base graphics</h2>
<p>The general idea for saving plots is open a graphics device, create the plot and then close the device. We will use <strong>png</strong> here. Check out <code>?png</code> for the arguments and other devices.</p>
<pre class="r"><code>dfr &lt;- data.frame(a=sample(1:100,10),b=sample(1:100,10))

png(filename=&quot;plot-base.png&quot;)
plot(dfr$a,dfr$b)
dev.off()</code></pre>
</div>
<div id="ggplot2" class="section level2">
<h2><span class="header-section-number">10.4</span> ggplot2</h2>
<p>The same idea can be applied to ggplot2, but in a slightly different way. First save the file to a variable, and then export the plot.</p>
<pre class="r"><code>p &lt;- ggplot(dfr,aes(a,b)) + geom_point()

png(filename=&quot;plot-ggplot-1.png&quot;)
print(p)
dev.off()</code></pre>
<p><strong>ggplot2</strong> also has another easier helper function to export images.</p>
<pre class="r"><code>ggsave(filename=&quot;plot-ggplot-2.png&quot;,plot=p)</code></pre>
</div>
</div>
<div id="getting-help" class="section level1">
<h1><span class="header-section-number">11</span> Getting help</h1>
<ul>
<li>Use <code>?function</code> to get function documentation</li>
<li>Use <code>??bla</code> to search for a function</li>
<li>Use <code>args(function)</code> to get the arguments to a function</li>
<li>Go to the package CRAN page/webpage for vignettes</li>
<li><a href="http://www.cookbook-r.com/">R Cookbook</a>: General purpose reference.</li>
<li><a href="https://www.statmethods.net/">Quick R</a>: General purpose reference.</li>
<li><a href="https://stackoverflow.com/">Stackoverflow</a>: Online community to find solutions to your problems.</li>
</ul>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">12</span> Session info</h1>
<pre><code>## R version 4.0.5 (2021-03-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.6 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] stringr_1.4.0               scales_1.1.1               
##  [3] fgsea_1.16.0                enrichR_3.0                
##  [5] rafalib_1.0.0               pvclust_2.2-0              
##  [7] pheatmap_1.0.12             biomaRt_2.46.3             
##  [9] edgeR_3.32.1                limma_3.46.0               
## [11] DESeq2_1.30.1               SummarizedExperiment_1.20.0
## [13] Biobase_2.50.0              MatrixGenerics_1.2.1       
## [15] matrixStats_0.61.0          GenomicRanges_1.42.0       
## [17] GenomeInfoDb_1.26.7         IRanges_2.24.1             
## [19] S4Vectors_0.28.1            BiocGenerics_0.36.1        
## [21] ggplot2_3.3.5               formattable_0.2.1          
## [23] kableExtra_1.3.4            dplyr_1.0.7                
## [25] lubridate_1.8.0             yaml_2.2.1                 
## [27] fontawesome_0.2.2.9000      captioner_2.2.3            
## [29] bookdown_0.24               knitr_1.36                 
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-152           bitops_1.0-7           bit64_4.0.5           
##  [4] progress_1.2.2         webshot_0.5.2          RColorBrewer_1.1-2    
##  [7] httr_1.4.2             tools_4.0.5            utf8_1.2.2            
## [10] R6_2.5.1               DBI_1.1.1              mgcv_1.8-34           
## [13] colorspace_2.0-2       withr_2.4.2            gridExtra_2.3         
## [16] prettyunits_1.1.1      tidyselect_1.1.1       curl_4.3.2            
## [19] bit_4.0.4              compiler_4.0.5         rvest_1.0.2           
## [22] xml2_1.3.2             DelayedArray_0.16.3    labeling_0.4.2        
## [25] genefilter_1.72.1      askpass_1.1            rappdirs_0.3.3        
## [28] systemfonts_1.0.3      digest_0.6.28          rmarkdown_2.11        
## [31] svglite_2.0.0          XVector_0.30.0         pkgconfig_2.0.3       
## [34] htmltools_0.5.2        dbplyr_2.1.1           fastmap_1.1.0         
## [37] highr_0.9              htmlwidgets_1.5.4      rlang_0.4.12          
## [40] rstudioapi_0.13        RSQLite_2.2.8          farver_2.1.0          
## [43] jquerylib_0.1.4        generics_0.1.1         BiocParallel_1.24.1   
## [46] RCurl_1.98-1.5         magrittr_2.0.1         GenomeInfoDbData_1.2.4
## [49] Matrix_1.3-2           Rcpp_1.0.7             munsell_0.5.0         
## [52] fansi_0.5.0            lifecycle_1.0.1        stringi_1.7.5         
## [55] zlibbioc_1.36.0        BiocFileCache_1.14.0   grid_4.0.5            
## [58] blob_1.2.2             crayon_1.4.1           lattice_0.20-41       
## [61] splines_4.0.5          annotate_1.68.0        hms_1.1.1             
## [64] locfit_1.5-9.4         pillar_1.6.4           rjson_0.2.20          
## [67] geneplotter_1.68.0     fastmatch_1.1-3        XML_3.99-0.8          
## [70] glue_1.4.2             evaluate_0.14          data.table_1.14.2     
## [73] vctrs_0.3.8            openssl_1.4.5          gtable_0.3.0          
## [76] purrr_0.3.4            tidyr_1.1.4            assertthat_0.2.1      
## [79] cachem_1.0.6           xfun_0.27              xtable_1.8-4          
## [82] survival_3.2-10        viridisLite_0.4.0      tibble_3.1.5          
## [85] AnnotationDbi_1.52.0   memoise_2.0.0          ellipsis_0.3.2</code></pre>
<hr />
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">

<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p style="color:#bdbdbd;" class="small">
<b>2021</b> • <a href="https://nbis.se/">NBIS</a> • <a href="https://www.scilifelab.se/">SciLifeLab</a>
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
