<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Content</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">





</div>


<!-- rmd lab header -->
<p><br></p>
<p>This page contains links to different lectures (slides) and practical exercises (labs) that are part of this workshop. The links below are similar to that under [link removed] but organised here by topic.</p>
<div class="boxy boxy-lightbulb">
<p>Input code blocks are displayed like shown below. The code language is displayed above the block. Shell scripts (<strong>SH</strong>) are to be executed in the linux terminal. <strong>R</strong> scripts are to be run in R either through the terminal, RGui or RStudio.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>command</code></pre>
</div>
<p><i class="fas fa-exclamation-circle"></i> Note   <i class="fas fa-lightbulb"></i> Tip   <i class="fas fa-comments"></i> Discuss   <i class="fas fa-clipboard-list"></i> Task</p>
</div>
<p><br></p>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>A brief introduction to the world of RNA and RNA-Seq technology.</p>
<ul>
<li><a href="slide_rna_intro.pdf" target="_blank">Introduction to RNA (Slide)</a><br />
</li>
<li><a href="slide_rnaseq_intro.pdf" target="_blank">Introduction to RNA-Seq (Slide)</a></li>
</ul>
<p>A primer to working on the unix/linux command line and working on a remote computing cluster Uppmax.</p>
<ul>
<li><a href="slide_uppmax_intro.pdf" target="_blank">Introduction to Linux &amp; Uppmax (Slide)</a></li>
<li><a href="lab_uppmax_intro.html" target="_blank">Working on Uppmax (Lab)</a></li>
</ul>
<p>RNA-Seq analyses is usually carried out in R programming language and it will be useful to learn some basic R.</p>
<ul>
<li><a href="lab_r.html" target="_blank">Introduction to R (Lab)</a></li>
</ul>
<p>This topic covers retrieving supporting data needed for RNA-seq analyses. These include gene annotation IDs such as mapping between Ensembl IDs and Gene IDs, GO terms and transcript IDs. We also cover retrieving genomic data from Ensembl.</p>
<ul>
<li><a href="lab_download.html" target="_blank">Downloading data (Lab)</a></li>
</ul>
</div>
<div id="main-lab" class="section level1">
<h1>Main lab</h1>
<div id="data" class="section level2">
<h2>Data</h2>
<p>In most of the exercises, we will use RNA-seq data (Illumina short reads) from the dataset <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE131032" target="_blank">GSE131032</a> <a href="https://www.nature.com/articles/s41467-019-10769-x" target="_blank">Czarnewski et al (2019) <em>Nat Comm</em></a>.</p>
<p><em>To elucidated through an unbiased manner which genes and pathways are differentially regulated during mouse colonic inflammation followed by a tissue regeneration phase. In particular, we took advantage of the widely used dextran sodium sulfate (DSS)-induced model of colitis. This model is one of the few characterized by a phase of damage followed by a phase of regeneration. Therefore, this model gave the possibility to identify also sets of genes essential in the regeneration phase, a key step towards the resolution of the inflammation. In short, mice were exposed to DSS in the drinking water for 7 days, then allowed to recover for the following 7 days. During this period, we collected colonic tissue samples every second day to then be analyzed by RNA sequencing (RNA-seq). Next, we performed a RNA-seq analysis from colonic samples throughout the experiment and computed differentially expressed genes (DEGs) taking the complete kinetics of expression into consideration for p-value estimation using EdgeR.</em></p>
<p>For this course, downsampled FASTQ files from this dataset from 2 experimental groups (day00 and day07, 3 samples each) will be used in the labs on read mapping, transcript assembly, visualization, quality control and differential expression. There are many relevant questions that could be asked based on these measurements, from several quality checks to understanding biological insignts.</p>
</div>
<div id="quality-control" class="section level2">
<h2>Quality control</h2>
<p>Before doing any other analysis on mapped RNA-seq reads it is always important to do quality control of your mapped reads and that you do not have any obvious errors in your RNA-seq data.</p>
<ul>
<li><a href="slide_after_sequencing_qc.pdf" target="_blank">QC of raw reads (Slide)</a></li>
<li><a href="lab_qc_beforemapping.html" target="_blank">Short read quality control (Lab)</a></li>
</ul>
</div>
<div id="mapping" class="section level2">
<h2>Mapping</h2>
<p>This section contains information on how to map reads to a reference genome using splice-aware aligner STAR and HISAT2.</p>
<ul>
<li><a href="slide_rnaseq_aligners.pdf" target="_blank">RNA-Seq aligners (Slide)</a></li>
<li><a href="lab_mapping.html" target="_blank">Mapping reads using STAR (Lab)</a></li>
</ul>
</div>
<div id="post-alignment-qc" class="section level2">
<h2>Post-alignment QC</h2>
<p>After alignment, the BAM files are inspected for various alignment metrics. Some of these include the number of reads mapped, number of unmapped reads, regions in the reference that reads map to, gene body coverage, signs of DNA contamination etc.</p>
<ul>
<li><a href="slide_after_mapping_qc.pdf" target="_blank">QC of alignments (Slide)</a></li>
<li><a href="lab_qc_aftermapping.html" target="_blank">Post-alignment quality control (Lab)</a></li>
</ul>
<p>BAM files are optionally visualised using integrated genome viewer.</p>
<ul>
<li><a href="lab_igv.html" target="_blank">Using IGV (Lab)</a></li>
</ul>
</div>
<div id="quantification" class="section level2">
<h2>Quantification</h2>
<p>Gene counts are quantified from BAM files using featureCounts.</p>
<ul>
<li><a href="slide_quantification.pdf" target="_blank">Quantification (Slide)</a></li>
<li><a href="lab_quantification.html" target="_blank">Quantification (Lab)</a></li>
</ul>
</div>
<div id="filtering-normalisation" class="section level2">
<h2>Filtering &amp; Normalisation</h2>
<ul>
<li><a href="slide_preprocessing.html" target="_blank">Filtering &amp; Normalisation (Slide)</a></li>
<li><a href="lab_preprocessing.html" target="_blank">Filtering &amp; Normalisation (Lab)</a></li>
</ul>
</div>
<div id="exploratory-data-analyses" class="section level2">
<h2>Exploratory data analyses</h2>
<p>Before commencing any quantitative analyses, it is important to run some exploratory analyses to access similarity between samples. This is a vital step to identify mislabelled samples, poor-quality samples and/or replicates that differ considerably. This section dives deeper into exploratory analyses PCA and hierarchical clustering.</p>
<ul>
<li><a href="slide_pca.pdf" target="_blank">Principal component analysis (Slide)</a></li>
<li><a href="slide_clustering.pdf" target="_blank">Cluster analysis (Slide)</a></li>
<li><a href="lab_eda.html" target="_blank">Exploratory data analyses (Lab)</a></li>
</ul>
</div>
<div id="differential-gene-expression" class="section level2">
<h2>Differential gene expression</h2>
<p>We find genes that are differentially expressed between our time points.</p>
<ul>
<li><a href="slide_dge.html" target="_blank">Differential gene expression (Slide)</a><br />
</li>
<li><a href="lab_dge.html" target="_blank">DGE using DESeq2 (Lab)</a></li>
</ul>
</div>
<div id="functional-analysis" class="section level2">
<h2>Functional analysis</h2>
<p>We will perform functional analysis on the differentially expressed genes to place them into a function context and possibly explain the biological consequences of DE. Methods covered are GSA (Gene set analysis) and GSEA (Gene set enrichment analysis).</p>
<ul>
<li><a href="slide_functional.pdf" target="_blank">Functional analyses (Slide)</a><br />
</li>
<li><a href="lab_functional.html" target="_blank">Functional analysis (Lab)</a></li>
</ul>
</div>
</div>
<div id="bonus-labs" class="section level1">
<h1>Bonus labs</h1>
<div id="pseudoaligners" class="section level2">
<h2>Pseudoaligners</h2>
<p>This is an alternative step INSTEAD of mapping, PA-QC and quantification. Kallisto uses FastQ reads and a reference transcriptome (cDNA+ncRNA) to quantify transcripts using rapid pseudo-alignment along with bootstrap replicates to assess quantification inaccuracy. Kallisto is significantly faster than STAR or HISAT2 and has a small memory footprint. Kallisto generates transcript counts. Differential transcript expression is carried out using Sleuth which utilises bootstrap replicates.</p>
<ul>
<li><a href="slide_pseudoaligners.pdf" target="_blank">Pseudoaligners (Slide)</a></li>
<li><a href="lab_kallisto.html" target="_blank">Mapping and quantification using Kallisto (Lab)</a></li>
</ul>
</div>
<div id="small-rna-analyses" class="section level2">
<h2>small RNA analyses</h2>
<p>RNA-seq differential analyses workflow on microRNAs from Fruit fly.</p>
<ul>
<li><a href="lab_smallrna.html" target="_blank">Small RNA-seq analyses (Lab)</a></li>
</ul>
</div>
<div id="assembly-annotation" class="section level2">
<h2>Assembly &amp; annotation</h2>
<p>Raw sequencing short reads are assembled into transcripts using two approaches. Genome-guided assembly using HiSat2 and StringTie. De-novo transcriptome assembly using Trinity. Assembled transcriptomes are functionally annotated to identify genes.</p>
<ul>
<li><a href="lab_assembly.html" target="_blank">Transcriptome assembly (Lab)</a><br />
</li>
<li><a href="lab_annotation.html" target="_blank">Transcriptome annotation (Lab)</a></li>
</ul>
<hr />
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
