<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Online Repositories</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">






</div>


<!-- rmd lab header -->
<p><br></p>
<div class="boxy boxy-exclamation boxy-yellow">
<p>Create a directory named <code>data</code> in your current working directory for input and output files.</p>
</div>
<p>Load R packages.</p>
<pre class="r"><code># BiocManager::install(&quot;GEOquery&quot;)
library(GEOquery)
library(Biobase)
library(rafalib)
rafalib::mypar(mar=c(6,2.5,2.5,1)) #sets nice arrangement for the whole document

# source download function
source(&quot;https://raw.githubusercontent.com/NBISweden/workshop-RNAseq/master/assets/scripts.R&quot;)</code></pre>
<div id="loading-a-dataset-from-geo-using-geoquery" class="section level2">
<h2><span class="header-section-number">0.1</span> Loading a dataset from GEO using GEOquery</h2>
<p>For this exercise, let’s try getting the full count matrix from the original <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE131032" target="_blank">GSE131032</a> dataset.</p>
<p>In some cases the simple <code>getGEO</code> will download the data, metadata and gene annotation into a ExpressionSet format, where you can then extract the information about the gene counts, metadata and gene annotations.</p>
<pre class="r"><code>#Get the GSE object from it
gset &lt;- GEOquery::getGEO(&quot;GSE131032&quot;)[[1]]
gset

#Get the count matrix from the GSE objetc
data &lt;- gset@assayData$exprs
dim(data)

#Get the sample metadata / phenotypic from the GSE object 
metadata &lt;- gset@phenoData@data
dim(metadata)

#Get the gene annotation data from the GSE object 
annot &lt;- gset@featureData@data
dim(annot)</code></pre>
<p>However, in most cases (in our experience) that is not the case since every dataset is a bit unique. So we recommend checking the files deposited individually and download what is most relevant for you. We can first list the files deposited:</p>
<pre class="r"><code>file_list &lt;- GEOquery::getGEOSuppFiles(&quot;GSE131032&quot;,fetch_files = F)
print(file_list)</code></pre>
<p>You can see the <code>RAW.tar</code> file, which contains the original count file deposited individually for every sample (under the GSM). The command below will dowlaod the file into a folder with the GSE name. This is a <code>.tar</code> compressed file that can be extracted.</p>
<pre class="r"><code>GEOquery::getGEOSuppFiles(&quot;GSE131032&quot;,fetch_files = T,filter_regex = &quot;_RAW.tar&quot;)
list.files(&quot;GSE131032&quot;)

#extract tar using bash (within R, but can be done outside too)
system(&quot;cd GSE131032; tar -xvf GSE131032_RAW.tar&quot;)
system(&quot;cd GSE131032; rm GSE131032_RAW.tar&quot;)

list.files(&quot;GSE131032&quot;)

system(&quot;cd GSE131032; gunzip *.gz&quot;)</code></pre>
<p>Having the individual files, we can combine them using a for loop or apply function:</p>
<pre class="r"><code>#list the files
list.files(&quot;GSE131032&quot;)

#Check the column names from the 1st file
head( read.delim(paste0(&quot;GSE131032/&quot;,list.files(&quot;GSE131032&quot;)[1])) )

#Do an apply function to get all the estimated counts from each file and compile into a single matrix
data &lt;- sapply(list.files(&quot;GSE131032&quot;,pattern = &quot;.tsv&quot;), function(x){
  x &lt;- read.delim(paste0(&quot;GSE131032/&quot;,x))[,&quot;est_counts&quot;]
  return(x)
})
rownames(data)
colnames(data)

#Add the gene names and sample names
rownames(data) &lt;- read.delim(paste0(&quot;GSE131032/&quot;,list.files(&quot;GSE131032&quot;)[1]))[,&quot;target_id&quot;]
colnames(data) &lt;- sub(&quot;_.*&quot;,&quot;&quot;,colnames(data))

#Check the column names from the 1st file
rownames(metadata) == colnames(data)

#Rename data
colnames(data) &lt;- metadata$title
head(data)</code></pre>
<p>Alternativelly, in the <code>file_list</code> above we could also simply load the count matrix.</p>
<pre class="r"><code>#list files available
print(file_list)

#Download  and extract the full matrix
GEOquery::getGEOSuppFiles(&quot;GSE131032&quot;,fetch_files = T,filter_regex = &quot;GSE131032_kallisto_counts.csv.gz&quot;)
system(&quot;cd GSE131032; gunzip GSE131032_kallisto_counts.csv.gz&quot;)

#Read the file (note that the sample names here are not the GSM accesccion, but the investigator&#39;s IDs)
data &lt;- read.delim(paste0(&quot;GSE131032/GSE131032_kallisto_counts.csv&quot;),sep = &quot;,&quot;,row.names = 1)
colnames(data)

#Those sample IDs are stored in one of the metadata columns
metadata$description
metadata$description == colnames(data)

#Rename data
colnames(data) &lt;- metadata$title
head(data)</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
