<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Quality control</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">






</div>


<!-- rmd lab header -->
<p><br></p>
<p>These steps are to be carried out after the fastq files have been aligned to the reference.</p>
<div id="mapper-log-files" class="section level1">
<h1><span class="header-section-number">1</span> Mapper log files</h1>
<p>The first step after you have finished your mapping is to get a general feel of how the mapping went. Most mapping programs produce some sort of summary output, either to a file or to standard out. For example, if using the mapper Bowtie you need to pipe that output to a file to see the summary statistics.</p>
<p>The most important parts to look at are the proportion of uniquely mapping, multi-mapping and unmapped reads. We ideally want the uniquely mapping reads to be as high as possible. Multi-mapping or unmapped reads could indicate poor quality of the reads, adapter contamination or other reasons for low quality scores.</p>
<div id="multiqc-logs-summary" class="section level2">
<h2><span class="header-section-number">1.1</span> MultiQC logs summary</h2>
<p>After mapping with star of all samples, we ran MultiQC to summarize all the log-files. In this case we had a folder structure with <strong>sampleName.hisat2_summary.txt</strong>, so to make sure that MultiQC understands what is the sample name, we used the <code>-dd 2</code> command (e.g. it splits up the path and names the samples after the second last part).</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>
multiqc -d -dd 2 .</code></pre>
</div>
</div>
</div>
<div id="rseqc" class="section level1">
<h1><span class="header-section-number">2</span> RSeQC</h1>
<p>The RSeQC package is one of many tools to get basic mapping statistics from your BAM files. This package provides a number of useful modules that can comprehensively evaluate high throughput sequence data, especially RNA-seq data. Some basic modules quickly inspect sequence quality, nucleotide composition bias, PCR bias and GC bias, while RNA-seq specific modules evaluate sequencing saturation, mapped reads distribution, coverage uniformity, strand specificity, etc. You can read more about the package at <a href="http://rseqc.sourceforge.net/" target="_blank">the RSeQC website</a>.</p>
<p>The RSeQC package contains many steps that are equivalent to FastQC analysis, e.g. read quality, sequence composition (NVC), GC-bias etc, but the results may be different since many of the low quality reads may not map to the genome and therefore will not be included in the BAM file.</p>
<p>Some steps of the RSeQC package require a file with gene annotations in BED format. These can be downloaded from sources such as UCSC, RefSeq and Ensembl. In this case, the RSeQC team have already created annotation files in some common formats that can be downloaded from their website, but if you have data for a less studied organism you may need to create a BED-file on your own.</p>
<p>Two annotation files have already been downloaded into subdirectory <code>references/annotations</code> for you to use. These are: <strong>Mus_musculus.GRCm38.101.HouseKeepingGenes.bed</strong> and <strong>Mus_musculus.GRCm38.101.bed</strong>.</p>
<p>In this tutorial, we will not run all the different parts of the RSeQC package, only the most relevant ones for this experiment. The different scripts in the RSeQC package are well described on <a href="http://rseqc.sourceforge.net/" target="_blank">their website</a>, so read the instructions there and specify the input/output files to fit your file names and folder structure.</p>
<p>The steps that we are going to run are:</p>
<ol style="list-style-type: decimal">
<li>geneBody_coverage.py</li>
<li>infer_experiment.py</li>
<li>junction_saturation.py</li>
<li>read_distribution.py</li>
</ol>
<p><i class="fas fa-exclamation-circle"></i> The <strong>geneBody_coverage.py</strong> script takes a very long time to run, so we have created a subsection of annotations to run it on. Use the file <code>Mus_musculus.GRCm38.101.HouseKeepingGenes.bed</code>. If you want to run multiple files this might take to long also. Then you can even further reduce the file by only using the first 500 house keeping genes. For more details look at code below how to do that and how to run all files in a loop.</p>
<p><i class="fas fa-lightbulb"></i> When running <strong>read_distribution.py</strong>, an out file cannot be specified. Instead you need to pipe (<code>&gt;</code>) the output to a file, or look at the output in the terminal.</p>
<p><i class="fas fa-clipboard-list"></i> Run RSeQC for one sample and have a look at your output.</p>
<ul>
<li>Do most of your reads map to genes?</li>
<li>Are the RNAseq data stranded or unstranded and if so on what strand is the read coming from?</li>
<li>Do you have even coverage along the genes?</li>
<li>Do the reads cover most splice junctions?</li>
</ul>
<div id="creating-bed12-file" class="section level2">
<h2><span class="header-section-number">2.1</span> Creating Bed12 file</h2>
<p><strong>Optional</strong></p>
<p>If you are working on a organism that does not have a annotation file in bed12 that is needed to use this RSeQC it is possible to transform a GTF file to Bed12 file. This is done by using tools from uscs and needs a extra step to first go from GTF to genePred and then from genePred to bed. <em>gtfToGenePred</em> and <em>genePredToBed</em> can both be downloaded from ucsc utilities homepage.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>
#To unzip all annotation files if you don&#39;t have
gunzip references/annotations/*.gz

 gtfToGenePred \
   references/annotations/Mus_musculus.GRCm38.101.gtf \
   references/annotations/Mus_musculus.GRCm38.101.genePred

 genePredToBed \
   references/annotations/Mus_musculus.GRCm38.101.genePred\
   references/annotations/Mus_musculus.GRCm38.101.bed
</code></pre>
</div>
</div>
<div id="multiqc-summary" class="section level2">
<h2><span class="header-section-number">2.2</span> MultiQC summary</h2>
<p>MutliQC can also include the information from RSeQC. Rerun MultiQC after you have done your RSeQC steps.</p>
<p>It was created using commands:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>multiqc -f -d -dd 2 .
</code></pre>
</div>
<p><i class="fas fa-comments"></i> Have a look at the reports. What is your conclusion, do your samples look good? Is there anything that looks strange in any sample, or do you feel comfortable using all the samples in your analysis?</p>
<hr />
</div>
</div>
<div id="code" class="section level1">
<h1><span class="header-section-number">3</span> Code</h1>
This is all that is needed to run all the steps above on data on course data.
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>

##########################################
###  RUN RSEQC scripts on bamfiles     ###
#########################################

mkdir RSeQCresults
mkdir RSeQCresults/geneBody_coverage

head -n 500 references/annotations/Mus_musculus.GRCm38.101.HouseKeepingGenes.bed &gt; references/annotations/Mus_musculus.GRCm38.101.HouseKeepingGenes.first500.bed

geneBody_coverage.py -r references/annotations/Mus_musculus.GRCm38.101.HouseKeepingGenes.first500.bed \
  -i hisat2_results/  -o RSeQCresults/geneBody_coverage/allSamples


 mkdir RSeQCresults/read_distribution
 mkdir RSeQCresults/infer_experiment
 mkdir RSeQCresults/junction_saturation
 mkdir RSeQCresults/junction_annotation



for i in $(ls hisat2_results/*.sorted.bam); do

 sample=${i%.sorted.bam}
 sample=${sample#hisat2_results/}


 ECHO &quot;running read distribution on $sample&quot;

 read_distribution.py  -i $i \
 -r references/annotations/Mus_musculus.GRCm38.101.bed \
 &gt;RSeQCresults/read_distribution/$sample.readCoverage.txt

 infer_experiment.py  -i $i  \
 -r references/annotations/Mus_musculus.GRCm38.101.bed \
 &gt;RSeQCresults/infer_experiment/$sample.infer_experiment.txt

junction_saturation.py -i $i  \
-r references/annotations/Mus_musculus.GRCm38.101.bed \
 -o RSeQCresults/junction_saturation/$sample

junction_annotation.py -i $i  \
-r references/annotations/Mus_musculus.GRCm38.101.bed \
-o RSeQCresults/junction_annotation/$sample


done


multiqc -f -d -dd 2 .
</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
