<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Kallisto</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">






</div>


<!-- rmd lab header -->
<p><br></p>
<div id="kallisto" class="section level1">
<h1><span class="header-section-number">1</span> Kallisto</h1>
<p><strong>Kallisto</strong> is an “alignment-free” RNA-Seq quantification method that runs very fast with a small memory footprint, so that it can be run on most laptops. It is a command-line program that can be downloaded as binary executables for Linux or Mac, or in source code format. For a first insight into the program, read <a href="https://liorpachter.wordpress.com/2015/05/10/near-optimal-rna-seq-quantification-with-kallisto/" target="_blank">here</a> and for the published article, see <a href="http://www.nature.com/nbt/journal/vaop/ncurrent/full/nbt.3519.html" target="_blank">here</a>.</p>
<p>Kallisto is geared towards quantification on the transcript (isoform) level, rather than the gene-level (although the latter can also be done by post-processing Kallisto output.) However, read assignment to transcript isoforms cannot (in general) be done unambiguously, so there is an intrinsic “quantification noise” or variability in this process. Kallisto can thus be run either in a single step (which is very fast) or in “bootstrap” mode (which takes longer, but can be done on several processors in parallel) in order to get uncertainty estimates for the expression levels - a kind of error bars for the quantification process. Running with bootstraps is mandatory if you want to perform differential expression analysis of isoforms with Sleuth (see below).</p>
<p>Kallisto is primarily meant for quantification of an existing set of FASTA sequences, that is, it does not perform transcript assembly and it cannot quantify the expression of novel transcripts that are not in the transcript index that you provide to it. With that said, you can of course use contigs from an assembly that you have produced in some other program in your Kallisto index. It would also be possible to use the software for eg: metagenomics or metatranscriptomics quantification.</p>
<div id="build-index" class="section level2">
<h2><span class="header-section-number">1.1</span> Build index</h2>
<p>Download the cDNA reference and create a index file to run on (This has already been done!). But if you want to do it yourself you can uncomment the code and run it yourselves.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#mkdir references/ref_kallisto
   
#cd references/ref_kallisto

#wget ftp://ftp.ensembl.org/pub/release-101/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz



#kallisto index \
#  --index kallisto_mm10_transcriptome_index.idx \
#  Mus_musculus.GRCm38.cdna.all.fa.gz


# cd ../..
</code></pre>
</div>
<p>It takes less than 10 minutes.</p>
</div>
<div id="pseudoalign-reads" class="section level2">
<h2><span class="header-section-number">1.2</span> Pseudoalign reads</h2>
<p>Now you can pseudoalign all your reads to the Kalisto index using Kalisto quantification</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>

  mkdir kallisto_quantification

 for sample_name in $(ls fastq/*.clean.fq.gz.1M_reads.fq.gz); do

    file=${sample_name#fastq/}
    sample=${file%.clean.fq.gz.1M_reads.fq.gz}
   
   echo $sample

     
    mkdir kallisto_quantification/$sample

    kallisto quant \
      --index references/ref_kallisto/kallisto_mm10_transcriptome_index.idx \
      --output-dir kallisto_quantification/$sample \
      --threads 3 \
      --single \
      --fragment-length 200 \
      --sd 20 \
     $sample_name

  done
</code></pre>
</div>
</div>
</div>
<div id="quantification" class="section level1">
<h1><span class="header-section-number">2</span> Quantification</h1>
<div id="get-gene-quantifications" class="section level2">
<h2><span class="header-section-number">2.1</span> Get gene quantifications</h2>
<p>Since Kalisto uses transcripts estimates and many programs, like DEseq and EdgeR, works better with gene counts there are programs to get gene counts from transctipt estimates in our case we will use txImport. TxImport can import estimates from a lot of different sources. You can read more about it in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html" target="_blank">txImport vignette</a></p>
<p>For the program to work it needs a two column file that assigns each transcript to a gene. This can found using ensembl database or other public databases. In our case the relationship between the transcripts and gene is found in the sequencenames in the cDNA file that we downloaded. By using awk we can extract the transcript gene dependency and write it to a file.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code># Example of a sequence name in file 
#  &gt;ENSMUST00000177564.1 cdna chromosome:GRCm38:14:54122226:54122241:1 gene:ENSMUSG00000096176.1 gene_biotype:TR_D_gene transcript_biotype:TR_D_gene gene_symbol:Trdd2 description:T cell receptor delta diversity 2 [Source:MGI Symbol;Acc:MGI:4439546]


# Extract all transcriptnames (1st) and genenames (4th) from  sequence names and write to a file.   
gzcat references/ref_kallisto/Mus_musculus.GRCm38.cdna.all.fa.gz| \
grep &#39;&gt;&#39; |
awk &#39;{FS= &quot; &quot;}BEGIN{ print &quot;TXNAME,GENEID&quot;};{print substr($1,2) &quot;,&quot; substr($4,6)};&#39; \
&gt;references/ref_kallisto/tx2gene.mm.GRCm38.cdna.csv 
</code></pre>
</div>
</div>
<div id="tximport" class="section level2">
<h2><span class="header-section-number">2.2</span> TxImport</h2>
<p>The next steps are carried out in R so make sure you start R with the correct packages.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>library(dplyr) # data wrangling
library(ggplot2) # plotting
library(DESeq2) # rna-seq
library(edgeR) # rna-seq
library(tximport) # importing kalisto transcript counts to geneLevels
library(readr) # Fast readr of files.
library(rhdf5) # read/convert kalisto output files.  </code></pre>
</div>
If <em>tximport</em> and <em>rhdf5</em> are not installed you can install them now.
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>#if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
#  install.packages(&quot;BiocManager&quot;)

#BiocManager::install(&quot;tximport&quot;)
#BiocManager::install(&quot;rhdf5&quot;)</code></pre>
</div>
Load the metadata for the samples.
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code># source download function
source(&quot;https://raw.githubusercontent.com/NBISweden/workshop-RNAseq/master/assets/scripts.R&quot;)

# download metadata

download_data(&quot;data/metadata_raw.csv&quot;)
mr &lt;- read.csv(&quot;data/metadata_raw.csv&quot;,header=TRUE,stringsAsFactors=F,row.names=1)

mr</code></pre>
</div>
<div id="convert-to-gene-counts" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Convert to gene counts</h3>
<p>Read in the Kalisto transcript abundances and and convert the TxImport.</p>
<p>TxImport is created to</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>setwd(&quot;~/RNAseq&quot;)
files &lt;- paste(&quot;kallisto_quantification&quot; ,  
               list.files(path = &quot;kallisto_quantification&quot;,pattern = &quot;abundance.tsv&quot;, recursive = TRUE),
               sep = &quot;/&quot;)
names(files) &lt;- mr$SampleName


tx2gene &lt;- read_csv(&quot;references/ref_kallisto/tx2gene.mm.GRCm38.cdna.csv&quot;)


txi.kallisto.tsv &lt;- tximport(files, type = &quot;kallisto&quot;, tx2gene = tx2gene, ignoreAfterBar = TRUE)</code></pre>
</div>
</div>
<div id="convert-to-deseq2-object-for-analysis" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Convert to DEseq2 object for analysis</h3>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>mr = mr %&gt;% mutate(Day = as.factor(Day))
dds &lt;- DESeqDataSetFromTximport(txi.kallisto.tsv, mr, ~Day)


test = DESeq(dds)</code></pre>
</div>
<p>and now you can go on with the DE similair as the one you did with the feature count table.</p>
</div>
<div id="convert-to-a-edger-object-for-analysis" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Convert to a edgeR object for analysis</h3>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>cts &lt;- txi.kallisto.tsv$counts
normMat &lt;- txi.kallisto.tsv$length
normMat &lt;- normMat/exp(rowMeans(log(normMat)))
library(edgeR)
o &lt;- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))
y &lt;- DGEList(cts)
y$offset &lt;- t(t(log(normMat)) + o)

# y is now ready for estimate dispersion functions see edgeR User&#39;s Guide</code></pre>
</div>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">3</span> Session info</h1>
<pre><code>## R version 4.0.5 (2021-03-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] scales_1.1.1                fgsea_1.16.0               
##  [3] enrichR_3.0                 rafalib_1.0.0              
##  [5] pvclust_2.2-0               pheatmap_1.0.12            
##  [7] biomaRt_2.46.3              edgeR_3.32.1               
##  [9] limma_3.46.0                DESeq2_1.30.1              
## [11] SummarizedExperiment_1.20.0 Biobase_2.50.0             
## [13] MatrixGenerics_1.2.1        matrixStats_0.60.1         
## [15] GenomicRanges_1.42.0        GenomeInfoDb_1.26.7        
## [17] IRanges_2.24.1              S4Vectors_0.28.1           
## [19] BiocGenerics_0.36.1         ggplot2_3.3.5              
## [21] formattable_0.2.1           kableExtra_1.3.4           
## [23] dplyr_1.0.7                 lubridate_1.7.10           
## [25] leaflet_2.0.4.1             yaml_2.2.1                 
## [27] fontawesome_0.2.2.9000      captioner_2.2.3            
## [29] bookdown_0.23               knitr_1.33                 
## 
## loaded via a namespace (and not attached):
##  [1] colorspace_2.0-2       rjson_0.2.20           ellipsis_0.3.2        
##  [4] XVector_0.30.0         rstudioapi_0.13        farver_2.1.0          
##  [7] bit64_4.0.5            AnnotationDbi_1.52.0   fansi_0.5.0           
## [10] xml2_1.3.2             splines_4.0.5          cachem_1.0.6          
## [13] geneplotter_1.68.0     jsonlite_1.7.2         annotate_1.68.0       
## [16] dbplyr_2.1.1           compiler_4.0.5         httr_1.4.2            
## [19] assertthat_0.2.1       Matrix_1.3-2           fastmap_1.1.0         
## [22] htmltools_0.5.2        prettyunits_1.1.1      tools_4.0.5           
## [25] gtable_0.3.0           glue_1.4.2             GenomeInfoDbData_1.2.4
## [28] rappdirs_0.3.3         fastmatch_1.1-3        Rcpp_1.0.7            
## [31] vctrs_0.3.8            svglite_2.0.0          nlme_3.1-152          
## [34] crosstalk_1.1.1        xfun_0.25              stringr_1.4.0         
## [37] rvest_1.0.1            lifecycle_1.0.0        XML_3.99-0.7          
## [40] zlibbioc_1.36.0        hms_1.1.0              RColorBrewer_1.1-2    
## [43] curl_4.3.2             gridExtra_2.3          memoise_2.0.0         
## [46] stringi_1.7.4          RSQLite_2.2.8          highr_0.9             
## [49] genefilter_1.72.1      BiocParallel_1.24.1    rlang_0.4.11          
## [52] pkgconfig_2.0.3        systemfonts_1.0.2      bitops_1.0-7          
## [55] evaluate_0.14          lattice_0.20-41        purrr_0.3.4           
## [58] htmlwidgets_1.5.3      labeling_0.4.2         bit_4.0.4             
## [61] tidyselect_1.1.1       magrittr_2.0.1         R6_2.5.1              
## [64] generics_0.1.0         DelayedArray_0.16.3    DBI_1.1.1             
## [67] pillar_1.6.2           withr_2.4.2            mgcv_1.8-34           
## [70] survival_3.2-10        RCurl_1.98-1.4         tibble_3.1.4          
## [73] crayon_1.4.1           utf8_1.2.2             BiocFileCache_1.14.0  
## [76] rmarkdown_2.10         progress_1.2.2         locfit_1.5-9.4        
## [79] grid_4.0.5             data.table_1.14.0      blob_1.2.2            
## [82] digest_0.6.27          webshot_0.5.2          xtable_1.8-4          
## [85] tidyr_1.1.3            openssl_1.4.4          munsell_0.5.0         
## [88] viridisLite_0.4.0      askpass_1.1</code></pre>
<hr />
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
