<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Transcriptome assembly</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">






</div>


<!-- rmd lab header -->
<p><br></p>
<div id="preparation" class="section level1">
<h1><span class="header-section-number">1</span> Preparation</h1>
<p>This exercise is meant to get you acquainted with the type of data you would normally encounter in an assembly/annotation project. We will for all exercises use data for the fruit fly (<em>Drosophila melanogaster</em>), as that is one of the currently best annotated organisms and there is plenty of high quality data available.</p>
<p>You can create the folders where you want but I would suggest a folder organisation, if you do not follow this organisation remember to put the correct path to your data.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>mkdir bonus/assembly
cd bonus/assembly</code></pre>
</div>
<p>Let’s link the raw fastq files.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>mkdir raw
cd raw
ln -s /proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/raw_computes/* .</code></pre>
</div>
<p>We will clone this Github repo from NBIS annotation team which will be used during this practical.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>git clone https://github.com/NBISweden/GAAS.git</code></pre>
</div>
<div id="assembling-transcripts-based-on-rna-seq-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Assembling transcripts based on RNA-seq data</h2>
<p>RNA-Seq data is in general very useful in annotation projects as the data usually comes from the actual organism you are studying and thus avoids the danger of introducing errors caused by differences in gene structure between your study organism and other species.</p>
<p>Important points to remember before starting working with RNA-Seq:</p>
<ul>
<li><p>Check if sequence data is paired-end or single-end. Last generation of sequenced short reads (since 2013) are almost all paired. Anyway, it is important to check that information, which will be useful for the tools used in the next steps.</p></li>
<li><p>Check if the sequence data are stranded. This information will be useful for the tools used in the next steps. (It is recommended to use stranded data to avoid transcript fusion during the transcript assembly process. That gives more reliable results.)</p></li>
<li><p>Left / L / forward / 1 have identical meaning. It is the same for Right / R /Reverse / 2</p></li>
</ul>
<div id="qc" class="section level3">
<h3><span class="header-section-number">1.1.1</span> QC</h3>
<p>We run a quick QC to check encoding version and fastq quality score format. We have to use FastQC tool.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load FastQC/0.11.5

mkdir fastqc
fastqc ./raw/ERR305399_1.fastq.gz -o fastqc/</code></pre>
</div>
<p>Copy the FastQC report HTML file to your system and inspect the results. You can use <code>SCP</code> to copy the file. Run the command below from a <strong>LOCAL</strong> terminal.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>scp user@rackham.uppmax.uu.se:[path-to-report.html] .</code></pre>
</div>
<p>Next, we check the fastq quality score format. For this, we will be using the scripts libraries available in the git gaas you need first to export the libraries and run the script <code>fastq_guessMyFormat.pl</code>.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>export PERL5LIB=$PERL5LIB:./GAAS/annotation/
./GAAS/annotation/Tools/bin/fastq_guessMyFormat.pl -i ./raw/ERR305399_1.fastq.gz</code></pre>
</div>
<p>In the normal mode, it differentiates between Sanger/Illumina1.8+ and Solexa/Illumina1.3+/Illumina1.5+. In the advanced mode, it will try to pinpoint exactly which scoring system is used.</p>
<p>More test can be made and should be made on RNA-seq data before doing the assembly, we have not time to do all of them during this course. have a look <a href="https://en.wikipedia.org/wiki/List_of_RNA-Seq_bioinformatics_tools" target="_blank">here</a></p>
</div>
</div>
</div>
<div id="guided-assembly" class="section level1">
<h1><span class="header-section-number">2</span> Guided assembly</h1>
<p>In this section, we run a genome-guided transcriptome assembly. We use Trimmomatic to trim the reads, HISAT2 to map reads to reference genome and StringTie to assemble reads into transcripts.</p>
<div id="trimmomatic" class="section level2">
<h2><span class="header-section-number">2.1</span> Trimmomatic</h2>
<p><a href="http://www.usadellab.org/cms/?page=trimmomatic" target="_blank">Trimmomatic</a> performs a variety of useful trimming tasks for illumina paired-end and single ended data.The selection of trimming steps and their associated parameters are supplied on the command line.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>mkdir trimmomatic
module load bioinfo-tools
module load trimmomatic/0.36</code></pre>
</div>
<p>The following command line will perform the following:
* Remove adapters (ILLUMINACLIP:TruSeq3-PE.fa:2:30:10)
* Remove leading low quality or N bases (below quality 3) (LEADING:3)
* Remove trailing low quality or N bases (below quality 3) (TRAILING:3)
* Scan the read with a 4-base wide sliding window, cutting when the average quality per base drops below 15 (SLIDINGWINDOW:4:15)
* Drop reads below the 36 bases long (MINLEN:36)</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>java -jar /sw/apps/bioinfo/trimmomatic/0.36/milou/trimmomatic-0.36.jar PE -threads 5 -phred33 ./raw/ERR305399_1.fastq.gz ./raw/ERR305399_2.fastq.gz trimmomatic/ERR305399.left_paired.fastq.gz trimmomatic/ERR305399.left_unpaired.fastq.gz trimmomatic/ERR305399.right_paired.fastq.gz trimmomatic/ERR305399.right_unpaired.fastq.gz ILLUMINACLIP:/sw/apps/bioinfo/trimmomatic/0.36/milou/adapters/TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36</code></pre>
</div>
</div>
<div id="hisat2" class="section level2">
<h2><span class="header-section-number">2.2</span> HISAT2</h2>
<p>Once the reads have been trimmed, we use <a href="https://ccb.jhu.edu/software/hisat2/index.shtml" target="_blank">HISAT2</a> to align the RNA-seq reads to a genome in order to identify exon-exon splice junctions.
HISAT2 is a fast and sensitive alignment program for mapping next-generation sequencing reads (whole-genome, transcriptome, and exome sequencing data) against a reference genome.</p>
<p>First you need to build an index of your chromosome 4</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>mkdir index

module load HISAT2
module load samtools/1.8
hisat2-build /proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/chromosome/chr4.fa ./index/chr4_index</code></pre>
</div>
<p>Then you can run HISAT2 :
* <code>--phred33</code> Input qualities are ASCII chars equal to the Phred quality plus 33. This is also called the “Phred+33” encoding, which is used by the very latest Illumina pipelines (you checked it with the script <em>fastq_guessMyFormat.pl</em>)<br />
* <code>--rna-strandness &lt;string&gt;</code> For single-end reads, use <strong>F</strong> or <strong>R</strong>. <strong>F</strong> means a read corresponds to a transcript. <strong>R</strong> means a read corresponds to the reverse complemented counterpart of a transcript. For paired-end reads, use either <strong>FR</strong> or <strong>RF</strong>. (<strong>RF</strong> means fr-firststrand see <a href="https://github.com/NBISweden/GAAS/blob/master/annotation/CheatSheet/rnaseq_library_types.md" target="_blank">here</a> for more explanation).<br />
* <code>--novel-splicesite-outfile &lt;path&gt;</code> In this mode, HISAT2 reports a list of splice sites in the file :
chromosome name <tab> genomic position of the flanking base on the left side of an intron <tab> genomic position of the flanking base on the right <tab> strand (+, -, and .) ‘.’ indicates an unknown strand for non-canonical splice sites.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>hisat2 --phred33 --rna-strandness RF --novel-splicesite-outfile hisat2/splicesite.txt -S hisat2/accepted_hits.sam -p 5 -x index/chr4_index -1 trimmomatic/ERR305399.left_paired.fastq.gz -2 trimmomatic/ERR305399.right_paired.fastq.gz</code></pre>
</div>
<p>Finally you need to change the SAM into BAM file and to sort it in order for StringTie to use the bam file to assemble the read into transcripts.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>samtools view -bS -o hisat2/accepted_hits.bam hisat2/accepted_hits.sam
samtools sort -o hisat2/accepted_hits.sorted.bam hisat2/accepted_hits.bam</code></pre>
</div>
</div>
<div id="stringtie" class="section level2">
<h2><span class="header-section-number">2.3</span> StringTie</h2>
<p><a href="https://ccb.jhu.edu/software/stringtie/" target="_blank">StringTie</a> is a fast and highly efficient assembler of RNA-Seq alignments into potential transcripts. It uses a novel network flow algorithm as well as an optional de novo assembly step to assemble and quantitate full-length transcripts representing multiple splice variants for each gene locus.
You can add as input an annotation from GTF/GFF3 file to calculate TPM and FPKM values.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load StringTie

stringtie hisat2/accepted_hits.sorted.bam -o stringtie/transcripts.gtf</code></pre>
</div>
<p>When done you can find your results in the directory <strong>outdir</strong>. The file <code>transcripts.gtf</code> includes your assembled transcripts.</p>
<p>You could now also visualise all this information using a genome browser, such as IGV. IGV requires a genome fasta file and any number of annotation files in GTF or GFF3 format (note that GFF3 formatted file tend to look a bit weird in IGV sometimes).</p>
<p>Transfer the GTF files to your computer using SCP:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>scp user@rackham.uppmax.uu.se:[path-to-transcripts.gtf] .</code></pre>
</div>
<p><i class="fas fa-comments"></i> Looking at your results, are you happy with the default values of StringTie (which we used in this exercise) or is there something you would like to change?</p>
</div>
</div>
<div id="de-novo-assembly" class="section level1">
<h1><span class="header-section-number">3</span> De-novo assembly</h1>
<p><a href="https://github.com/trinityrnaseq/trinityrnaseq/wiki" target="_blank">Trinity</a> assemblies can be used as complementary evidence, particularly when trying to polish a gene build with Pasa. Before you start, check how big the raw read data is that you wish to assemble to avoid unreasonably long run times.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load trinity/2.4.0
module load samtools

Trinity --seqType fq --max_memory 32G --left ./raw/ERR305399_1.fastq.gz --right ./raw/ERR305399_2.fastq.gz --CPU 5 --output trinity --SS_lib_type RF</code></pre>
</div>
<p>Trinity takes a long time to run (like hours), you can stop the program when you start it and have a look at the results, look in <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/RNAseq/trinity</code> and the output is <code>Trinity.fasta</code>.</p>
<p>In order to compare the output of StringTie and the output of Trinity we need to map the Trinity transcript to the chr4 of Drosophila.</p>
<p>We’ll use the GMAP software to align the Trinity transcripts to our reference genome. Trinity contains a utility that facilitates running GMAP, which first builds an index for the target genome followed by running the gmap aligner:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>module load gmap-gsnap

mkdir gmap

/sw/apps/bioinfo/trinity/2.4.0/rackham/util/misc/process_GMAP_alignments_gff3_chimeras_ok.pl --genome /proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/chromosome/chr4.fa --transcripts /proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/RNAseq/trinity/Trinity.fasta &gt; gmap/transcript_trinity.gff</code></pre>
</div>
<div id="assembly-quality" class="section level2">
<h2><span class="header-section-number">3.1</span> Assembly quality</h2>
<p>There are different ways of assessing the quality of your assembly, you will find some of them <a href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/Transcriptome-Assembly-Quality-Assessment" target="_blank">here</a>.</p>
<p>We will run BUSCO to check the the quality of the assembly. <a href="https://busco.ezlab.org/" target="_blank">BUSCO</a> provides measures for quantitative assessment of genome assembly, gene set, and transcriptome completeness (what we are going to do here). Genes that make up the BUSCO sets for each major lineage are selected from orthologous groups with genes present as single-copy orthologs in at least 90% of the species in the chosen branch of tree of life.</p>
<p>For the Trinity results :</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>module load BUSCO/3.0.2b

/sw/apps/bioinfo/BUSCO/3.0.2b/rackham/bin/run_BUSCO.py -i /proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/RNAseq/trinity/Trinity.fasta -o busco_trinity -l $BUSCO_LINEAGE_SETS/arthropoda_odb9 -m tran -c 5</code></pre>
</div>
<p>Busco will take 30 to run so you can check the results in <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/RNAseq/busco_trinity</code></p>
<p>For the guided assembly results, you need first to extract the transcript sequences from the gtf transcript file :</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>module load BioPerl

export PERL5LIB=$PERL5LIB:./GAAS/annotation/
./GAAS/annotation/Tools/bin/gff3_sp_extract_sequences.pl --cdna -g transcripts.gtf -f /proj/uppstore2017171/courses/RNAseqWorkshop/downloads/assembly_annotation/chromosome/chr4.fa -o transcripts_stringtie.fa</code></pre>
</div>
<p>Then you can run BUSCO again :</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>/sw/apps/bioinfo/BUSCO/3.0.2b/rackham/bin/run_BUSCO.py -i stringtie/transcripts_stringtie.fa -o busco_stringtie -l $BUSCO_LINEAGE_SETS/arthropoda_odb9 -m tran -c 5</code></pre>
</div>
<p><i class="fas fa-clipboard-list"></i> Compare the two BUSCO results, what do you think happened for StringTie?</p>
</div>
<div id="whats-next" class="section level2">
<h2><span class="header-section-number">3.2</span> What’s next?</h2>
<p>Now you are ready either to annotate your RNA-Seq or you can use then to do the genome annotation.</p>
<p>For the de-novo assembly you can use the <code>Trinity.fasta</code> file obtained.
For the genome-guided assembly you can either use the StringTie results <code>transcripts.gtf</code>, but you will often need to reformat it into a gff file.</p>
<hr />
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
