<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Mapping</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:22px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Mapping</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>

</div>


<!-- rmd lab header -->
<p><br></p>
<p>In this tutorial we will use a splice aware aligner HISAT2 that is the best solution when using a laptop or a personal computer with less than 16GB. If you have a computer with memory &gt;32GB or access to computer cluster it is worthwhile to consider using STAR and compare the results.</p>
<div id="hisat2" class="section level1">
<h1><span class="header-section-number">1</span> HISAT2</h1>
<div id="data" class="section level2">
<h2><span class="header-section-number">1.1</span> Data</h2>
<p>All FASTQ files that you will need for this exercise can be found in the <code>fastq</code> subfolder of your data folder.</p>
<p>A pre-built mouse genome index for HISAT2 can be found in the <code>references/hisat2_build</code> subfolder of your data folder.</p>
</div>
<div id="mapping" class="section level2">
<h2><span class="header-section-number">1.2</span> Mapping</h2>
<p>Here, you will map the reads to the mouse reference genome using the RNA-seq aligner <strong>HISAT2</strong>.</p>
<p>There are many features that can be tweaked using HISAT2. For more information on all flags that can be used go <a href="https://daehwankimlab.github.io/hisat2/manual/">here</a>.</p>
<p>Read below for the flags we use for this exercise. Remember to change filenames accordingly so that you can run your program properly and know which files you have used.</p>
<p>Now you can map the reads from one of the samples (or several; itâ€™s up to you which ones) using a command such as the one below.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>mkdir outDir

# for paired end reads
hisat2 -p N -x path/to/index/fileName \
  -1 path/to/reads/sample_1.fastq \
  -2 path/to/reads/sample_2.fastq \
  -S outDir/hisat2.sam \
  --summary-file outDir/hisat2_summary.txt


# for single end reads
hisat2 -p N  -x path/to/index/fileName \
  -U path/to/reads/sample.fastq \
  -S outDir/hisat2.sam \
  --summary-file outDir/hisat2_summary.txt
</code></pre>
</div>
<p>The flags used are:</p>
<ul>
<li><code>-p N</code> specifies the number of threads that will be used by the program</li>
<li><code>-x /path/to/index/fileName</code> specifies the path to the pre-built genome index. Note that the index consists of multiple files ending in <code>.ht2</code>, and only the shared part of the filename should be indicated (e.g.Â <code>genome</code> if the files are called <code>genome.1.ht2</code>, <code>genome.2.ht2</code>, etc).</li>
<li><code>-1 /path/to/reads/sample_1.fastq</code> is where you should list the first-read FASTQ file of paired end reads that you wish to map</li>
<li><code>-2 /path/to/reads/sample_2.fastq</code> is where you should list the second-read FASTQ file of paired end reads that you wish to map</li>
<li><code>-U /path/to/reads/sample.fastq</code> is where you should list the single end read FASTQ file that you wish to map</li>
<li><code>-S outDir/hisat2.sam</code> is the name of the result file that will be created</li>
<li><code>--summary-file outDir/hisat2_summary.txt</code> is the name of a file for summary information about the alignments</li>
</ul>
<p>This should run fairly quickly and create the files you specified with <code>-S</code> and <code>--summary-file</code>.</p>
<p>If everything worked, HISAT2 should report some statistics about how many reads were mapped, on your terminal and in the summary file.</p>
<p><i class="fas fa-comments"></i> <strong>Try to answer the following:</strong></p>
<ul>
<li>How many RNA-seq reads were provided as input to HISAT2?</li>
<li>How many of those reads were mapped by HISAT2?</li>
<li>How many reads were uniquely mapped, i.e.Â mapped to one genomic location?</li>
<li>In general, do the alignments seem to be good? I.e. do they cover the entire reads and contain few mismatches?</li>
</ul>
<p>To answer these questions, you should look at the input and output from HISAT2. You may also need to consult the <a href="https://daehwankimlab.github.io/hisat2/manual/">HISAT2 manual</a>, <a href="https://en.wikipedia.org/wiki/FASTQ_format">information about the FASTQ format</a> and the <a href="https://github.com/samtools/hts-specs">SAM format specification</a>.</p>
</div>
</div>
<div id="convert-sam-to-bam" class="section level1">
<h1><span class="header-section-number">2</span> Convert SAM to BAM</h1>
<p>If you were able to run HISAT2, this should have produced files with mapped reads in SAM format. These files need to be converted to <em>sorted</em> and <em>indexed</em> BAM files for efficient downstream analysis.</p>
<p>You should try to give the BAM files representable names, in order to make it easier to manage your files. A good naming scheme for BAM files is to use names that indicate what you mapped and how. As an example, if you mapped sample 12 using HISAT2 you could create a file named <code>sample12.mmusculus.HISAT2.bam</code>.</p>
<p>The most commonly used tool for converting from SAM to BAM is <a href="http://www.htslib.org/doc/samtools.html">Samtools</a> (follow the link for more information about Samtools).</p>
<p>The Samtools command to convert from SAM to a sorted BAM file is:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>samtools sort -o output.bam input.sam</code></pre>
</div>
<p>Remember to use an appropriate filename instead of <code>output.bam</code>! Next, we need to index the BAM file.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>samtools index properName.bam</code></pre>
</div>
<p>The indexing step should create an index file with the suffix <code>.bai</code>. This sorted, indexed BAM file can be viewed in the Integrative Genomics Viewer (IGV).</p>
<p>You can also get a report on your mapped reads using the samtools command <em>flagstat</em>:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>samtools flagstat properName.sorted.bam &gt; properName.sorted.flagstat</code></pre>
</div>
<p><i class="fas fa-lightbulb"></i> Since the BAM file contains all the information from the original SAM file, remember to remove the SAM file once you are finished, in order to free up disk space.</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>rm input.sam</code></pre>
</div>
<hr />
</div>
<div id="code-to-run-all-the-steps-above-on-data-on-course-data" class="section level1">
<h1><span class="header-section-number">3</span> Code to run all the steps above on data on course data</h1>
<div class="chunk-title-parent">
<div class="chunk-title small">
sh
</div>
<pre class="sh"><code>


###############################################################################
###          MAP READS TO REFERENCE                                       #####
###############################################################################


######################################
###  BUILD A HISAT2 REFERENCE     ####
######################################





###########################
###  ALIGN WITH HISAT2  ###
###########################
mkdir hisat2_results

#map all the small files
 for i in $(ls fastq/*.clean.fq.gz.1M_reads.fq.gz); do

    file=${i#fastq/}
    sample=${file%.clean.fq.gz.1M_reads.fq.gz}
   #Single-end read
   hisat2 -p 4 \
    -x references/hisat2_build/mm10 \
    -U $i \
    -S hisat2_results/$sample.1M.sam \
    --summary-file hisat2_results/$sample.hisat2_summary.txt


 done


#map  the large file
 for i in $(ls fastq/*.clean.fq.gz); do

    file=${i#fastq/}
    sample=${file%.clean.fq.gz}
   #Single-end read
   hisat2 -p 4 \
    -x references/hisat2_build/mm10 \
    -U $i \
    -S hisat2_results/$sample.10M.sam \
    --summary-file hisat2_results/$sample.10M.hisat2_summary.txt



 done




############################
####   CONVERT TO BAM   ####

############################


for i in $(ls hisat2_results/*.sam); do

    file=${i#hisat2_results/}
    sample=${file%.sam}
 
 
 # This is how it was done in a older version of samtools  (Version: 0.1.19-44428cd)
  #samtools view -bS -o hisat2_results/$sample.bam hisat2_results/$sample.sam 
  #samtools sort  hisat2_results/$sample.bam hisat2_results/$sample.sorted
  
 # This is how it should be done in the version that is used in our conda enviroment (Version: 1.10 (using htslib 1.10.2))
  samtools sort -o hisat2_results/$sample.sorted.bam hisat2_results/$sample.sam
  samtools index hisat2_results/$sample.sorted.bam
  samtools flagstat hisat2_results/$sample.sorted.bam &gt; $sample.sorted.flagstat

done

############################
####   REMOVE SAM FILES ####
############################


# Remove the intermediate sam and bamfiles.
for i in $(ls hisat2_results/*.sorted.bam); do

    file=${i#hisat2_results/}
    sample=${file%.sorted.bam}

 if test -f &quot;hisat2_results/$sample.sorted.bam&quot;; then
    echo &quot;$sample.sorted.bam exists. Removing $sample.sam and $sample.bam&quot;
    rm hisat2_results/$sample.sam

    # This intermediate step is no longer needed anymore 
    #rm hisat2_results/$sample.bam
 fi

done
</code></pre>
</div>
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">

<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p style="color:#bdbdbd;" class="small">
<b>2021</b> â€¢ <a href="https://nbis.se/">NBIS</a> â€¢ <a href="https://www.scilifelab.se/">SciLifeLab</a>
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
